---
title: "Paper"
author: "Sarah Killcoyne"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document:
    fig_height: 5
    fig_width: 5
    toc: yes
    toc_depth: 4
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
library(tidyverse)
library(broom)
library(BarrettsProgressionRisk)
library(ggrepel)
library(kableExtra)
library(reshape2)
library(gridExtra)
library(RColorBrewer)
library(ggfortify)

library(OACUtils)

source('lib/load_patient_metadata.R')
source('lib/common_plots.R')

dir = '~/Data/BarrettsProgressionRisk/'
info.dir = paste0(dir,'/QDNAseq')

patient.file = list.files(info.dir, pattern='All_patient_info.xlsx', recursive=T, full.names=T)
demo.file = list.files(info.dir, pattern='Demographics_full.xlsx', recursive=T, full.names=T)

if (length(patient.file) != 1 | length(demo.file) != 1)
  stop(paste("Missing/too many patient info file(s) in", info.dir))

all.patient.info = read.patient.info(patient.file, demo.file, set='All')
patient.info = bind_rows(all.patient.info$info, all.patient.info$replicates) # for the moment...

patient.info = arrange(patient.info, Status, Hospital.Research.ID, Endoscopy.Year, Pathology)
sum.patient.data = as_tibble(summarise.patient.info(patient.info))

patient.info = patient.info %>% mutate( Pathology = recode(Pathology, 'BE'='NDBE'))

```

# Demographics

```{r demo}
cohortSummary = sum.patient.data %>% dplyr::group_by(Status) %>% 
  dplyr::summarise(
  'Total Patients'=length(unique(Patient)),
  'Total Samples'=sum(total.samples),
  'Range follow-up (dates)'=paste(min(start.year), max(end.year), sep='-'),
  'Range follow-up (years)'=paste(range(end.year-start.year), collapse='-'),
  #'Mean follow-up (years)'=paste(round(mean(end.year-start.year), 1),'+/-',round(sd(end.year-start.year),1)),
  #'Total Range endoscopies'=paste(range(total.endos), collapse='-'),
  #'Mean Endoscopies'=paste(round(mean(total.endos), 1),'+/-',round(sd(total.endos),1)),
  'Median total endoscopies (IQR)'=paste0(median(total.endos), ' (', paste(quantile(total.endos, prob=c(0.25,0.75)), collapse='-'), ')'),
  'Median age at diagnosis (IQR)'=paste0(median(age.diagnosis), ' (', paste(quantile(age.diagnosis, prob=c(0.25,0.75)), collapse='-'), ')'),
  'Mean Length (not recorded)'=paste0(paste(round(mean(M,na.rm=T), 1),'+/-',round(sd(M,na.rm=T),1)), ' (', length( which(is.na(M) )),')'),
  'Gender F:M'=paste(table(gender),collapse=':'),
  'Smoking N:Y (not recorded)'=paste0(paste(table(smoking), collapse=':'), ' (', length(which(is.na(smoking) )), ')')
) %>% left_join(patient.info %>% group_by(Status, Pathology) %>% tally %>% spread(Pathology,n), by='Status')


cohortSummary = cohortSummary %>% add_row(
  Status = 'p-value (test)', 'Total Patients'= '', 'Total Samples' = '', 'Range follow-up (dates)' = '', 
  'Range follow-up (years)' = paste0(signif(do(sum.patient.data, tidy(wilcox.test(end.year-start.year ~ Status, data=., exact=F)))$p.value,3), ' (Wilcox)'),
  'Median age at diagnosis (IQR)' = paste0(signif(do(sum.patient.data, tidy(wilcox.test(age.diagnosis ~ Status, data = ., exact=F)))$p.value,3), ' (Wilcox)'),
  'Mean Length (not recorded)' = paste0(signif(do(sum.patient.data, tidy(wilcox.test(M ~ Status, data=., exact=F)))$p.value,3), ' (Wilcox)'),
  'Gender F:M' = paste0(signif(dplyr::summarise(sum.patient.data, pvalue = chisq.test(gender, Status)$p.value)$pvalue,3), ' (ChiSq)'),
  'Smoking N:Y (not recorded)' = paste0(signif(dplyr::summarise(sum.patient.data, pvalue = chisq.test(smoking, Status)$p.value)$pvalue,3), ' (ChiSq)')
)
  
cohortSummary %>% t %>% kable(caption='Training Cohort') %>% kable_styling()


fit = glm(Status ~ age.diagnosis+M+gender+initial.path, data=mutate(sum.patient.data, initial.path = factor(initial.path,ordered=F)), family = binomial(link='logit'))
tidy(fit) %>% kable(caption = 'Logistic model to test effects of (matched) demographic factors on progression') %>% 
  kable_styling(full_width = F)

sample.age = patient.info %>% group_by(Samplename, Status, Pathology) %>% dplyr::summarise(sample.age = 2020-Endoscopy.Year)

ggplot(sample.age, aes(Status, sample.age, group=Status, fill=Status)) + geom_jitter(width=0.2) + geom_boxplot(alpha =0.5, width=0.5, outlier.color = NA) + 
  scale_y_continuous(breaks = seq(0,28,5)) + labs(y='Years', title='Sample age') + plot.theme

ggplot(sample.age, aes(Pathology, sample.age, group=Pathology, fill=Pathology, color=Status)) + 
  geom_jitter(width=0.2) + geom_boxplot(width=0.5, outlier.color = NA, color='black') + 
  scale_y_continuous(breaks = seq(0,28,5)) + labs(y='Years', title='Sample age') + plot.theme

rm(cohortSummary, sample.age, fit)
```


```{r, echo=F, eval=F}
grid.arrange(
  ggplot(patient.info, aes(p53.Status, fill=Status)) + geom_bar(position = position_dodge()) + labs(title='p53 Staining') + plot.theme,
  tableGrob( 
    patient.info %>% mutate(p53.Status = forcats::fct_explicit_na(p53.Status)) %>% group_by(Status, p53.Status) %>% tally() %>% spread(p53.Status, n)),
  nrow=2, as.table=T, heights=c(4,1) )  
```

# Copy number data set up

To use GLMs across the patients I merge all samples from all patients and overlap the copy number segments. Where segments from a sample get split, the value of the original sample is retained for each of the split samples.

## y( Progressors (1) vs Non (0) )

The labels are split such that all samples from progressor patients are (1) and all samples from non-progressors are (0).

```{r labelsPNP, echo=F, message=F, include=T}
cache.dir = paste(dir, 'Analysis/models_5e6_all/50kb', sep='/') 
file = paste(cache.dir, 'model_data.Rdata', sep='/')
if (!file.exists(file))
  stop(paste("Missing data file", file))

load(file, verbose=F)
melted = melt(dysplasia.df)

p = ggplot(melted, aes(Var2, Var1, fill = value)) + geom_tile() +
  scale_fill_gradient2() + 
  theme(axis.text = element_blank()) + labs(x='Feature', y='sWGS Sample')
leg = get_legend(p)
ggsave('plots/samples.png', plot = p + theme(legend.position = 'none'), width=3, height=4, units='in')

file = paste(cache.dir, 'all.pt.alpha.Rdata', sep='/')
load(file, verbose=F)
rm(plots, coefs, performance.at.1se, models)

residuals = do.call(bind_rows, purrr::map(list.files(paste0(dir,'/Analysis/pcf_perPatient/50kb'), 'residuals', recursive = T, full.names = T), 
                              function(f) { read_tsv(f, col_types = 'ccddddl') }
))

rm(p,leg, melted)
```

We have `r table(labels)[1]` samples from non-progressors and `r table(labels)[2]` samples from progressors.

A few samples failed our QC process and were excluded from the model training process.
`r residuals %>% filter(!Pass) %>% kable(caption = 'Samples that failed post-QDNAseq QC') %>% kable_styling(full_width = F)`

Example subset of the data matrix:
`r dysplasia.df[1:5, 1:5] %>% kable(caption=paste("dimensions:", paste(dim(dysplasia.df), collapse=', '))) %>% kable_styling('striped')`

## Global copy number plots

```{r echo=F, warning=F, message=F, eval=T, fig.height=6, fig.width=10}
rm(residuals)

chr.lengths = BarrettsProgressionRisk:::chrInfo() %>% dplyr::filter(chr %in% c(1:22)) %>%
  dplyr::mutate(chrom = factor(sub('chr','', chrom), levels=c(1:22), ordered=T))

arm.loc = arrange(bind_rows(
  (chr.lengths %>% dplyr::group_by(chrom) %>% dplyr::mutate('start'=1, 'end'=chr.cent-cent.gap))[c('chrom','start','end')],
  (chr.lengths %>% dplyr::group_by(chrom) %>% dplyr::mutate('start'=chr.cent+cent.gap, 'end'=chr.length))[c('chrom','start','end')]), chrom)

globalCNdf = as_tibble( t(dysplasia.df[,grep('^\\d+:',colnames(dysplasia.df))]), rownames = 'locs') %>% 
  separate(locs, c('chr','start','end'), sep=':|-', remove=F) %>% 
  dplyr::mutate_at(vars(start, end), as.numeric) %>%
  dplyr::mutate( seg.type = ifelse( (start %in% arm.loc$start & end %in% arm.loc$end), 'arms','5MB'  ) ) %>%
  melt(id.vars=c('locs','chr','start','end', 'seg.type')) %>% as_tibble() %>%
  left_join(chr.lengths %>% dplyr::select(chrom,chr.length), by=c('chr'='chrom')) %>%
  left_join(patient.info %>% dplyr::select(matches('Hospital|Patient|Samplename|Pathology|Status')), by=c('variable'='Samplename')) %>%
  dplyr::mutate(Status = factor(recode(Status, NP='Non-Progressor',P='Progressor'), ordered=T)) %>% 
  dplyr::mutate( 
    cn = factor(unlist(purrr::map(.x=value, .f=get.cn.state)),  levels=c('loss','neutral','gain'), ordered = T),
    chr = factor(chr, levels = chr.lengths$chrom, ordered=T))

globalCNdf %<>% filter(!is.na(Status))

n.status <- c('Progressor'=paste0('P\n(n=',length(unique(subset(globalCNdf, Status == 'Progressor')$variable)), ')'),
              'Non-Progressor'=paste0('NP\n(n=',length(unique(subset(globalCNdf, Status == 'Non-Progressor')$variable)),')'))

globalCNdf %>% filter(seg.type != 'arms') %>% dplyr::select(Patient, Status, Pathology, variable, chr, start, end, value) %>% 
  dplyr::rename_all(list(~str_to_title(.))) %>% dplyr::rename('Chromosome' = 'Chr', 'Sample' = 'Variable', 'CN Value' = 'Value') %>%
  write_tsv('plots/global_cn_values.tsv')
  


globalCN.plot = cn.mtn.plot(globalCNdf %>% filter(seg.type != 'arms'), labeller(Status=n.status, chr=label_value))

cnLegend = get.legend(globalCN.plot)
globalCN.plot + theme(legend.position = 'none')

gp = globalCN.plot + labs(x='Chromosome', y='Relative CN', title='') + 
  theme(legend.position='none', axis.text.y=element_text(size=8), strip.text.x=element_text(size=8), strip.text.y=element_text(size=9), title=element_text(face = 'bold'), axis.title=element_text(face="plain")) + 
  scale_y_continuous(breaks = c(-20,0,20) )
ggsave(filename='plots/globalCN.png', plot=gp, width=10, height=2.2, units='in', dpi=300)


## NDBE
nb.status <- c('Progressor'=paste0('P\n(n=',length(unique(subset(globalCNdf, Status == 'Progressor' & Pathology == 'NDBE')$variable)), ')'),
              'Non-Progressor'=paste0('NP\n(n=',length(unique(subset(globalCNdf, Status == 'Non-Progressor' & Pathology == 'NDBE')$variable)),')'))

globalBECN.plot = cn.mtn.plot(filter(globalCNdf, seg.type != 'arms' & Pathology == 'NDBE'), labeller(Status=nb.status, chr=label_value)) + labs(title='NDBE samples')

gbe = globalBECN.plot + labs(y='Relative CN',x='Chromosome', title="") +
  theme(legend.position='none', axis.text.y=element_text(size=8), strip.text.x=element_text(size=8), strip.text.y=element_text(size=9), title=element_text(face = 'bold'), axis.title=element_text(face="plain")) + 
  scale_y_continuous(breaks = c(-10,0,10)) 
ggsave(filename='plots/globalBECN.png', plot=gbe, width=10, height=2.2, units='in', dpi=300)

rm(arm.loc, cnLegend)
```

## *Complexity Scores*

```{r loaddata, message=F, warning=F, echo=F}
file = paste(cache.dir, 'all.pt.alpha.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading", file))
  load(file, verbose=F)
} else {
  stop(paste("No model file", file, "found"))
}

cx = dysplasia.df[,'cx']
complexity.score = tibble('Status'=labels, 'cx'=cx[names(labels)], 'Samplename' = names(labels)) %>%
  mutate(Status = factor(Status, labels = c('Non-Progressor','Progressor'))) %>%
  left_join(patient.info[,c('Samplename','Pathology','Patient')],by=c('Samplename'))

tt = wilcox.test(subset(complexity.score, Status == 'Progressor')$cx, subset(complexity.score, Status == 'Non-Progressor')$cx)

cx.labels = complexity.score %>% group_by(Status) %>% tally() %>% spread(Status, n)

complexity.score %>% dplyr::rename('Complexity Score' = 'cx') %>% dplyr::select(-Patient, -Pathology) %>% write_tsv('plots/cx_score.tsv')
 
pv = ggplot(complexity.score, aes(Status, cx, fill=Status)) + 
  geom_jitter(width=0.2, color='grey39') + 
  geom_boxplot(outlier.colour = NA, alpha=0.8, width=0.3) +
  scale_x_discrete(breaks = unique(complexity.score$Status), 
                   labels = c('Non-Progressor' = paste0('Non-Progressor\n(n=',cx.labels$`Non-Progressor`,')'),'Progressor' = paste0('Progressor\n(n=',cx.labels$`Progressor`,')'))) + 
  labs(x='', y='cx', title='Per sample cx scores', subtitle=paste('p-value=', format.pval(tt$p.value, digits=2))) + theme(legend.position = 'none') +
  plot.theme
pv
ggsave('plots/per_sample_cx.png', plot=pv + theme(text=element_text(size=12)), width=4, height=4, units='in', dpi=300)

rocX = pROC::roc(Status ~ cx, data=complexity.score, auc=T, ci=T, of='thresholds')
roc.plot(rocX, title='Per sample cx') + plot.theme

ggsave('plots/per_sample_cx_roc.png', plot=roc.plot(rocX, title='Per sample cx') + plot.theme + theme(text=element_text(size=12)), width=4, height=4, units='in', dpi=300)

rm(cx,complexity.score, tt, cx.labels, pv, rocX)
```

# Building the model

We built the model using a 50-fold cross-validation process on the patients (rather than samples) for various lambda and alpha values to find the best fitting model. 

```{r model, message=F, warning=F, echo=F, fig.height=16, fig.width=16}
folds = 10; splits = 5 
file = paste(cache.dir, 'all.pt.alpha.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading", file))
  load(file, verbose=F)
} else {
  stop(paste("No model file", file, "found"))
}

do.call(grid.arrange, c(plots, top='All samples, 10fold, 5 splits'))
ggsave('plots/eyebrows.png', plot=do.call(grid.arrange, c(plots, top='All samples, 10fold, 5 splits')), width=10, height=10, units='in', dpi=300)
```

```{r echo=F, fig.height=5, fig.width=5}
autoplot(models$`0.9`) + theme(legend.position = 'none') + plot.theme
ggsave('plots/glm_0.9.png', plot=autoplot(models$`0.9`) + theme(legend.position = 'none') + plot.theme, width=5, height=5, units='in', dpi=300)
```


## Model performance

```{r model_perf, message=F, warning=F, echo=F, fig.width=6, fig.height=6}
classification.rate = as_tibble(matrix(ncol=5, nrow=0, dimnames=list(c(), c('model','mean','sme','stable.coef', 'all.coef'))))

mp = model.performance(performance.at.1se, coefs, folds, splits )
mp$plot = mp$plot + labs(title=paste('All samples,', folds, 'folds,', splits, 'patient splits'))
mp$plot
ggsave('plots/performance/ridge_v_lasso.png', plot=mp$plot, width=5, height=5, units='in', dpi=300)

select.alpha = as.character(0.9)
# Genomic regions only
most.stable = grep('^\\d+', names(which(mp$stable.coeffients[[select.alpha]] >= 0.75 )), value=T)
most.stable = cbind.data.frame(do.call(rbind, strsplit(most.stable, ':|-')), most.stable)
colnames(most.stable) = c('chr','start','stop', 'name')

most.stable$chr = as.integer(as.character(most.stable$chr))
most.stable = arrange(most.stable, chr, start)

classification.rate = add_row(classification.rate, 'model'='All samples', 'mean'=mp$performance[select.alpha, 'mean'], 'sme'=mp$performance[select.alpha, 'sme'],                              'stable.coef'=length(which(mp$stable.coeffients[[select.alpha]] >= 0.75)), 'all.coef'=length(mp$stable.coeffients[[select.alpha]]))

mp$performance[,c('n.Coef','25%','50%','75%', '100%')] %>% kable(caption='Number of features stable in >=n% of folds') %>% kable_styling(full_width = F)

rm(mp,most.stable)
```

## Feature Stability

## Remove HGD/IMC & LGD Samples and check the model

The initial model is trained against `r nrow(dysplasia.df)` samples with `r ncol(dysplasia.df)` features. The samples include the final (diagnostic) HGD/IMC samples from progressors. 

### Remove HGD/IMC

```{r noHGD, echo=F, message=F, warning=F, fig.height=16, fig.width=16, eval=T}
rm(performance.at.1se, coefs, mp)
file = paste(cache.dir, 'nohgd.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading file", file))
  load(file, verbose=F)
} else {
  stop(paste("No model file", file, "found"))
}

nohgd.coefs = coefs[[select.alpha]]
ggsave('plots/nohgd_eyebrow.png', plot=do.call(grid.arrange, c(no.hgd.plots, top="No HGD/IMC samples", ncol=2)), height=10, width=10, units='in', dpi=300)
```

```{r, echo=F, fig.width=6, fig.height=6}
mp = model.performance(performance.at.1se, coefs, folds, splits)
mp$plot = mp$plot + labs(title='No HGD/IMC samples')
mp$plot
ggsave('plots/performance/ridge_v_lasso_nohgd.png', plot=mp$plot, width=5, height=5, units='in', dpi=300)

classification.rate = add_row(classification.rate, 'model'='No HGD/IMC', 
                              'mean'=mp$performance[select.alpha, 'mean'], 'sme'=mp$performance[select.alpha, 'sme'],
                              'stable.coef'=length(which(mp$stable.coeffients[[select.alpha]] >= 0.75)), 'all.coef'=length(mp$stable.coeffients[[select.alpha]]))
rm(no.hgd.plots,performance.at.1se,coefs,mp)
```

### Remove LGD

```{r noLGD, echo=F, warning=F, message=F, fig.height=16, fig.width=16}
file = paste(cache.dir, 'nolgd.Rdata', sep='/')
if (file.exists(file)) {
  load(file, verbose=F)
} else {
  stop(paste("No model file", file, "found"))
}
```

```{r, echo=F, fig.width=6, fig.height=6}
mp = model.performance(performance.at.1se, coefs, folds, splits)
mp$plot = mp$plot + labs(title='No LGD/HGD/IMC samples')
mp$plot
ggsave('plots/performance/no_lgd_perf.png', plot=mp$plot, height=5, width=5, units='in', dpi=300)

classification.rate = add_row(classification.rate, 'model'='No LGD/HGD/IMC', 'mean'=mp$performance[select.alpha, 'mean'], 'sme'=mp$performance[select.alpha, 'sme'],
                              'stable.coef'=length(which(mp$stable.coeffients[[select.alpha]] >= 0.75)), 'all.coef'=length(mp$stable.coeffients[[select.alpha]]))
rm(nolgd.plots,performance.at.1se,coefs,mp)
```

## Overall

Even without dysplastic samples our model is well able to classify samples as either progressive or non.

```{r, echo=F, warning=F, message=F, fig.width=6, fig.height=6}
p = ggplot(classification.rate, aes(model, mean, group=model, fill=model)) + geom_col() +
  geom_text( aes(label=round(mean, 2)), vjust=2, size=5, color='gray39', show.legend=F) +
  geom_text( aes(y=0.6, label=paste(all.coef, " coef\n(", stable.coef, ")", sep='') ), color='gray39', size=5) +
  geom_errorbar(aes(ymin=mean-sme, ymax=mean+sme),color='grey39', width=0.1, size=1) + 
#  geom_point(size=2) +
  labs(x='', y='Classification rate', title=paste('Classification rate per model at', select.alpha)) + 
  plot.theme + theme(axis.text.x=element_text(angle=45,hjust=1,face="bold",size=12), legend.position = 'none') +
  ylim(0,1)
p
ggsave('plots/performance/compare_classification.png', plot=p, height=5, width=3, units='in', dpi=300 )
rm(p)
```

# Features

```{r, echo=F, warning=F, message=F, fig.height=5, fig.width=5}
file = paste(cache.dir, 'all.pt.alpha.Rdata', sep='/')
load(file, verbose=F)
rm(plots,performance.at.1se,models,cvs,labels)

featureStability = rowSums(coefs[[select.alpha]][,-1])/(splits*folds)
hazards = apply( exp( t(dysplasia.df[,rownames(coefs[[select.alpha]])]) * coefs[[select.alpha]][,1]), 1, function(x) {
  sd(x)/mean(x)
})

features = as_tibble(hazards, rownames = 'label') %>% dplyr::rename(hazards = value) %>% 
  dplyr::mutate(coef = coefs[[select.alpha]][,1], stability = featureStability) #%>% arrange(-hazards)

p = ggplot(features, aes(label,hazards, fill=coef)) + geom_col() + 
  scale_fill_gradient2() + coord_flip() + theme_classic() + 
  theme(axis.ticks.y = element_blank(), panel.grid = element_blank(), axis.text = element_blank(), legend.position = 'none') + labs(x='', y='', title='Effects') 
ggsave('plots/effects.png', plot=p, width = 3, height=1.5, units='in')
rm(p)
  
cvRR = features
qq = quantile(features$hazards, seq(0,1,.15))
p = ggplot(features, aes(coef, hazards, col=hazards >= qq[length(qq)])) + 
  geom_point(alpha=0.8, show.legend = F) + 
  geom_text_repel(aes(x=coef, label=ifelse(hazards >= qq[length(qq)], label, '')),show.legend=F) +
  scale_color_manual(values=c('firebrick3','darkblue')) +
  labs(title='Top features by cv(RR)', x='Coefficient value', y='cvRR') + plot.theme 
p
ggsave('plots/haz_coef.png', plot=p, height=5, width=5, units='in', dpi=300)
rm(qq,p)

cvRR %>% mutate(`Gain/Loss` = ifelse(coef > 0 ,'Gain', 'Loss')) %>% 
  dplyr::select(label, hazards, `Gain/Loss`) %>% arrange(desc(hazards)) %>%
  write_tsv('plots/coef_effect.tsv')
```

`r features %>% arrange(desc(hazards)) %>% kable(caption='Selected coefficients') %>% kable_styling(full_width=F)`

```{r, echo=F, warning=F, message=F, fig.height=10, fig.width=12}
features = features %>% filter(label != 'cx') %>% separate(label, c('chr','start','end'), sep=':|-', remove=F) %>% 
  dplyr::mutate_at(vars(start, end), list(as.numeric)) %>% 
  mutate(chr = factor(chr,levels=c(1:22))) %>% 
  dplyr::mutate(ymax = ifelse(coef > 0, max(globalCN.plot$data$mean.value), min(globalCN.plot$data$mean.value)), gl = factor(ifelse(coef > 0, 'gain','loss'))) %>%
  left_join(chr.lengths %>% dplyr::select(chr, chr.length), by=c('chr')) %>% mutate(chr = factor(chr, levels=c(1:22), ordered=T))

custom.max<-function(cf, hazards, ymax) {
  if (cf > 0) {
    max = hazards*ymax
    max = ifelse(max > ymax, ymax, max)
  } else {
    max = hazards*ymax
    max = ifelse(max < ymax, ymax, max)
  }
  return(max)
}
features = features %>% rowwise() %>% dplyr::mutate('max'=custom.max(coef, hazards, ymax)) %>% arrange(chr)

p = globalCN.plot + geom_rect(data=features %>% filter(gl == 'gain'), aes(xmin=start, xmax=end, ymin=0, ymax=max), color='goldenrod2', alpha=0, show.legend=F) + 
  geom_rect(data= features %>% filter(gl == 'loss'), aes(xmin=start, xmax=end, ymin=0, ymax=max), color='blue3', alpha=0, show.legend=F) + theme(legend.position = 'none')
ggsave('plots/global_cn_coefs.png', plot=p, width=16, height=4.6, units='in', dpi=300)
rm(p)

p = globalBECN.plot +  geom_rect(data=features %>% filter(gl == 'gain'), aes(xmin=start, xmax=end, ymin=0, ymax=max), color='goldenrod2', alpha=0, show.legend=F) + 
  geom_rect(data=features %>% filter(gl == 'loss'), aes(xmin=start, xmax=end, ymin=0, ymax=max), color='blue3', alpha=0, show.legend=F) + theme(legend.position = 'none')
ggsave('plots/ndbe_cn_coefs.png', plot=p, width=16, height=4.6, units='in', dpi=300)
rm(p)
```

# Nested leave-one-out predictions

Leave out each patient (both P and NP) and run a new CV fit each time then predict the samples for the patient that was left out.  Fitted model includes HGD samples.

```{r leaveoneout, echo=F, message=F, warning=F, fig.width=5, fig.height=5}
error = read_tsv('~/Data/BarrettsProgressionRisk/Analysis/pcf_perPatient/50kb/sample_error.tsv', col_types = 'cdD')

file = paste(cache.dir, 'loo_0.9.Rdata', sep='/')
if (file.exists(file)) {
  load(file, verbose=F)
} else {
  stop(paste("Missing model file", file))
}
loo.coefs = nzcoefs
rm(nzcoefs)

error = error %>% left_join(pg.samp %>% dplyr::select(Status, Samplename, RR), by=c('Sample'='Samplename')) %>% 
  dplyr::mutate(P.High = BarrettsProgressionRisk:::pi.hat(RR+Error), P.Low = BarrettsProgressionRisk:::pi.hat(RR-Error), Abs.Error = P.High-P.Low )

p = ggplot(dplyr::select(pg.samp, Status, Samplename, RR), aes(reorder(Samplename, -RR), 1, fill=RR)) + geom_tile() + 
  scale_fill_gradient2(midpoint = 5, low = scales::muted('blue'), high = scales::muted('red')) + 
  scale_y_continuous(expand=c(0,0)) + 
  coord_flip() + labs(y='', x='') + theme_bw() + 
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank(), legend.position = 'none') 
ggsave('plots/per_sample_RR.png', plot = p, width=0.5, height=3, units='in')
rm(p)

pg.samp = pg.samp %>% mutate(Pathology = recode(Pathology, 'BE' = 'NDBE')) %>%
  left_join(error %>% dplyr::select(-RR,-Status), by=c('Samplename'='Sample'))

pg.samp %>% dplyr::select(Status, -Pathology, Prediction) %>% dplyr::rename('Probability' = 'Prediction') %>%
  dplyr::mutate(Status = recode_factor(Status, 'NP' = 'Non-Progressor', 'P' = 'Progressor')) %>%
  write_tsv('plots/loo_all_samples.tsv')

performance.at.1se = enframe(performance.at.1se, value='performance.at.1se')#names(performance.at.1se) = pg.samp$Samplename
performance.at.1se$name = unique(pg.samp$Patient)
p = ggplot(performance.at.1se, aes(y=performance.at.1se, x='LOO')) + 
  geom_boxplot(fill='lightblue', color='darkgrey', outlier.fill=NA, outlier.color=NA, size=0.8) + geom_jitter(color='grey39', width=0.3) +
  #geom_label(data=filter(performance.at.1se %in% range(performance.at.1se)), aes(y=performance.at.1se, label=round(performance.at.1se, 3)), fontface='bold', fill='lightblue') +
  labs(y='Classification rate', x='LOO model', title='Performance for LOO', subtitle='All samples model') +
  plot.theme
p
ggsave('plots/performance_loo.png', plot=p, width=5, height=5, units='in', dpi=300)
rm(p,performance.at.1se,plots,coefs,fits)
```

### Prediction ROC curves

#### LOO minus HGD

```{r loonohgd, echo=F, message=F, warning=T}
file = paste(cache.dir, 'loonohgd.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading file", file))
  load(file, verbose=T)
  rm(plots,coefs,nzcoefs,fits)
} else {
  stop(paste("No file", file, "found"))
}

perf = cbind.data.frame(perf=performance.at.1se, pt=unique(pg.sampNOHGD$Patient))

p = ggplot(melt(perf, id.vars='pt'), aes(x=variable,y=value, group=variable, fill=variable)) +
    geom_boxplot(fill='lightcoral', color='darkgrey', outlier.fill=NA, outlier.color=NA, size=0.8) +
    geom_jitter(color='grey39', width=0.3) + plot.theme + 
    geom_label(data=melt(subset(perf, perf %in% range(perf) ), id.vars='pt'), aes(x=variable, y=value, label=round(value, 3)), fontface='bold', fill='lightcoral') +
    theme(legend.position = 'none') + labs(x='LOO model', y='Classification rate', title='Performance for LOO', subtitle='HGD samples excluded model')
ggsave('plots/performance/loo_perf_nohgd.png', plot=p, width=5, height=5, units='in', dpi=300)
rm(p, perf,performance.at.1se)

pg.sampNOHGD = pg.sampNOHGD %>% filter(!is.na(Prediction))

pg.sampNOHGD %>% dplyr::select(Status, -Pathology, Prediction) %>% dplyr::rename('Probability' = 'Prediction') %>%
  dplyr::mutate(Status = recode_factor(Status, 'NP' = 'Non-Progressor', 'P' = 'Progressor')) %>%
  write_tsv('plots/loo_noHGD_samples.tsv')

rocNOHGD = pROC::roc(Status ~ Prediction, data = pg.sampNOHGD[c('Status', 'Prediction')], ci=T, of='thresholds')
roc.plot(rocNOHGD, 'Model excluding HGD/IMC',incS = T, 0.3) 

pROC::coords(rocNOHGD, 0.3, transpose=F) %>% add_column('AUC'=pROC::auc(rocNOHGD)) %>% mutate_all(list(~round(.,2))) %>% 
  kable(caption = 'LOO excluding HGD/IMC') %>% kable_styling(full_width = F)

ggsave(filename='plots/auc/roc_nohgd_model.png', plot=roc.plot(rocNOHGD, 'Model excluding HGD/IMC',T, 0.3) , width=6, height=6, units='in', dpi=300)

rm(pg.sampNOHGD)
```


#### LOO minus HGD/LGD

```{r loonolgd, echo=F, message=F, warning=T}
file = paste(cache.dir, 'loonolgd.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading file", file))
  load(file, verbose=T)
  rm(plots,performance.at.1se,coefs,nzcoefs,fits)
} else {
  stop(paste("No file", file, "found"))
}

pg.sampNOLGD = pg.sampNOLGD %>% filter(!is.na(Prediction))

rocNOLGD = pROC::roc(Status ~ Prediction, data=pg.sampNOLGD[c('Status', 'Prediction')], ci=T, of='thresholds')
roc.plot(rocNOLGD, 'Model excluding LGD/HGD/IMC',T,0.3) 

pROC::coords(rocNOLGD, 0.3, transpose=F) %>% add_column('AUC'=pROC::auc(rocNOLGD)) %>% mutate_all(list(~round(.,2))) %>% 
  kable(caption = 'LOO excluding LGD/HGD/IMC') %>% kable_styling(full_width = F)

ggsave(filename='plots/auc/roc_nolgd_model.png', plot=roc.plot(rocNOLGD, 'Model excluding LGD/HGD/IMC',T, 0.3) , width=6, height=6, units='in', dpi=300)

rm(pg.sampNOLGD)
```

#### Insilico splits

```{r insilico_rocs, echo=F, message=F, warning=F, fig.height=8, fig.width=8}
load(file=paste(cache.dir, 'insilico_splits.Rdata', sep='/'), verbose=F)
```

```{r prediction_cutoff, echo=F, message=F, warning=F, fig.height=7, fig.width=7}
preds = pg.samp %>% dplyr::select(Status, Prediction)
roc = pROC::roc(Status ~ Prediction, data=preds, auc=T, ci=T, of='thresholds')

roc.plot(roc,'Discovery Cohort all samples ROC',incS=T)
round(pROC::coords(roc, 'best', transpose=F),2) %>% kable(caption = 'All samples ROC') %>% kable_styling(full_width = F)

ggsave(filename='plots/auc/training_auc.png', plot=roc.plot(roc,'Discovery Cohort all samples ROC',incS = T),width=6, height=6, units='in', dpi=300)
ggsave(filename='plots/auc/training_auc_supp.png', plot=roc.plot(roc,'Discovery Cohort all samples ROC'), width=6, height=6, units='in', dpi=300)

roc$model = 'CN Model'
roc$is = 'discovery'

insilico.rocs[['discovery']] = roc
df = do.call(rbind, lapply(insilico.rocs, function(r) 
    cbind.data.frame('AUC' = as.numeric(r$auc), 'Specificity'=rev(r$specificities), 'Sensitivity'=rev(r$sensitivities),'model'=r$model, 'is'=as.character(r$is) 
)))

i.thresholds = do.call(bind_rows, purrr::map(insilico.rocs, function(r) pROC::coords(r, 'best') %>% as_tibble %>% 
                                               mutate(AUC = r$auc, model=r$model, is=as.character(r$is)))) %>%
  rename_at(vars(matches('thresh|spec|sens')), funs(str_to_title)) 

blues = RColorBrewer::brewer.pal(5,'Greys')
p = ggplot(df, aes(Specificity, Sensitivity)) + 
    geom_line(color=blues[2]) + 
    geom_line(data=filter(df, is == 'discovery'), color=blues[5], size=1) +
    geom_line(data=df %>% filter( (AUC == max(AUC) ) & is != 'discovery'), color=blues[4], linetype = 'dotted', size=0.75) +
    geom_line(data=df %>% filter( (AUC == min(AUC) ) & is != 'discovery'), color=blues[4], linetype = 'dotted', size=0.75) +
    scale_x_reverse() +
    geom_label_repel(data=filter(i.thresholds, AUC %in% range(AUC) | is == 'discovery'), aes(label=paste0('FPR ', round(1-Specificity,2), ' TPR ', round(Sensitivity,2), '\nAUC ',round(AUC,2))), color='black') +
    #scale_color_brewer(palette='Blues') +
    labs(title='Discovery cohort: 50 models', subtitle='18:70 random patient splits', x='Specificity (1-FPR)', y='Sensitivity (TPR)')  + plot.theme +
    theme(legend.position = 'none', text = element_text(size=14))
p
ggsave(filename='plots/auc/splits_roc.png', plot=p, width=6, height=6, units='in', dpi=300)
rm(insilico.rocs, insilico.perf, blues, i.thresholds, df, p)
```

### p53 ROC curve

```{r p53roc, echo=F, fig.height=2, fig.width=2}
p53samples = subset(patient.info, !is.na(p53.Status), select=c('Samplename','p53.Status','Status', 'Pathology'))
p53roc = pROC::roc(as.numeric(p53samples$Status)-1, as.integer(p53samples$p53.Status)-1)

#pROC::ci.auc(p53roc$auc)
#pROC::coords(p53roc, 'best',transpose=T)

p53roc$model = 'p53 IHC'
roc.plot(p53roc, 'p53 IHC', incS = T) 
ggsave('plots/auc/p53_roc.png', plot=roc.plot(p53roc, 'p53 IHC', incS = T) , width=6, height=6, units="in", limitsize=F, dpi = 300)
```

### Pathology ROC curve

Based on HGD/IMC/LGD
```{r pathroc, echo=F, fig.height=4, fig.width=12 }
pathsamples = patient.info[,c('Pathology','Status')]
pathroc = pROC::roc((as.integer(pathsamples$Status)-1), ifelse(pathsamples$Pathology %in% c('LGD','HGD','IMC'), 1, 0) )
#pROC::ci.auc(pathroc$auc)
#pROC::coords(pathroc, 'best')

pathroc$model = 'HGD & LGD'
ggsave('plots/auc/dysplasia_auc.png', plot=roc.plot(pathroc, title='Dysplasia (HGD & LGD)', incS = T), width=6, height=6, units="in", limitsize=F, dpi = 300)
       
hgdpathroc = pROC::roc((as.integer(pathsamples$Status)-1), ifelse(pathsamples$Pathology %in% c('HGD','IMC'), 1, 0) )
hgdpathroc$model = 'HGD Only'
ggsave('plots/auc/hgd_auc.png', plot=roc.plot(hgdpathroc, title='HGD only', incS = T), width=6, height=6, units="in", limitsize=F, dpi = 300)

grid.arrange(roc.plot(pathroc, 'Pathology (LGD,HGD,IMC)'), roc.plot(hgdpathroc, 'Pathology (HGD,IMC)', incS = T), ncol=2)
```

### Per Endoscopy ROCs
```{r, echo=F, warning=F, message=F,fig.height=4, fig.width=4}

tmp = pg.samp %>% dplyr::select(Patient, Status, Endoscopy.Year, PID, Prediction) %>%
  arrange(Patient, Endoscopy.Year, PID) 

pout = tibble()
for (pt in unique(pg.samp$Patient)) {
  pout = pout %>% bind_rows( tmp %>% filter(Patient == pt) %>% group_by(Endoscopy.Year, PID) %>% dplyr::mutate(Endoscopy = dplyr::group_indices()) )
}
  
pout %>% dplyr::select(-Endoscopy.Year, -PID) %>% dplyr::rename('Probability' = 'Prediction') %>% 
  dplyr::mutate(Status = recode_factor(Status, 'NP'='Non-Progressor', 'P'='Progressor')) %>% 
  write_tsv('plots/aggregate_rocs.tsv')

pendo = pg.samp %>% dplyr::group_by(Patient, Status, Hospital.Research.ID, PID ) %>% 
  dplyr::summarise( 'max'=max(Prediction), 'avg'=mean(Prediction), 'n'=length(PID) )

rocEndoM = pROC::roc(Status ~ max, data=pendo, ci=T, of='thresholds')
rocEndoM$model = 'Max per Endoscopy'
ggsave('plots/auc/maxendo_auc.png', plot=roc.plot(rocEndoM, title='Max per Endoscopy', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

rocEndoA = pROC::roc(Status ~ avg, data=pendo, ci=T, of='thresholds')
rocEndoA$model = 'Mean per Endoscopy'
ggsave('plots/auc/avgendo_auc.png', plot=roc.plot(rocEndoA, title='Mean per Endoscopy', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

lastEndo = pg.samp %>% group_by(Patient, Hospital.Research.ID, Status,Endoscopy.Year, PID) %>% 
  arrange(desc(Endoscopy.Year), PID) %>% dplyr::summarise(avg = mean(Prediction) ) %>% ungroup %>% 
  group_by(Patient) %>% dplyr::filter( Endoscopy.Year == max(Endoscopy.Year) ) %>% dplyr::select(-Endoscopy.Year, -PID)

rocLastEndo = pROC::roc(Status ~ avg, data=lastEndo, ci=T, of='thresholds')
rocLastEndo$model = 'Last endoscopy'
ggsave('plots/auc/lastendo_auc.png', plot=roc.plot(rocLastEndo, title='Last endoscopy', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

firstEndo = pg.samp %>% group_by(Patient, Hospital.Research.ID, Status,Endoscopy.Year, PID) %>% 
  arrange(Endoscopy.Year, PID)  %>% dplyr::summarise(avg = mean(Prediction) ) %>% ungroup %>% 
  group_by(Patient) %>% dplyr::filter( Endoscopy.Year == min(Endoscopy.Year) ) %>% dplyr::select(-Endoscopy.Year, -PID)

rocFirstEndo = pROC::roc(Status ~ avg, data=firstEndo, ci=T, of='thresholds')
rocFirstEndo$model = 'First endoscopy'
ggsave('plots/auc/firstendo_auc.png', plot=roc.plot(rocFirstEndo, title='First Endoscopy', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

pendo.rand = pg.samp %>% dplyr::group_by(Patient, Status, PID) %>%
  dplyr::summarise( r.Prob = sample(Prediction, 1), n = length(PID) ) %>% filter(n > 1)
rocRandEndo = pROC::roc(Status ~ r.Prob, data=pendo.rand, ci=T, of='thresholds')
rocRandEndo$model = 'Per Endoscopy Random'
ggsave('plots/auc/randomendo_auc.png', plot=roc.plot(rocRandEndo, title='Random per Endoscopy', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

# endo.thresholds = bind_cols('roc'=c('Max','Mean','Last','First','Random'),
#   bind_rows(
#     pROC::coords(rocEndoM, 'best', transpose = T),
#     pROC::coords(rocEndoA, 'best', transpose = T),
#     pROC::coords(rocLastEndo, 'best', transpose = T),
#     pROC::coords(rocFirstEndo, 'best', transpose = T),
#     pROC::coords(rocRandEndo, 'best', transpose = T)
#   ))

multi.roc.plot(list(rocEndoM, rocEndoA, rocLastEndo, rocFirstEndo, rocRandEndo), threshold = 0.3) + labs(title='Endoscopy prediction ROCs')

rm(pendo, pendo.rand)
```

### Per Patient ROC
```{r echo=F, fig.height=4, fig.width=12}
# Excluding the final endoscopy
ptendo = do.call(bind_rows, lapply(unique(pg.samp$Hospital.Research.ID), function(pt) {
  x = pg.samp %>% filter(Hospital.Research.ID == pt)  
  x %>% filter(Endoscopy.Year != max(x$Endoscopy.Year))
})) %>% group_by(Patient, Hospital.Research.ID, Status) %>% summarise_at(vars(Prediction), funs(max, mean), na.rm=T)

ptendo = ptendo %>% filter(!is.na(mean))

rocPtM = pROC::roc(Status ~ max, data=ptendo, ci=T, of='thresholds')
rocPtM$model = 'Max per Patient'
#pROC::coords(rocPtM, 'best', transpose=T)
ggsave('plots/auc/maxpt_auc.png', plot=roc.plot(rocPtM, title='Max per Patient (excl. final)', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

rocPtA = pROC::roc(Status ~ mean, data=ptendo, ci=T, of='thresholds')
rocPtA$model = 'Mean per Patient'
#pROC::coords(rocPtA, 'best', transpose=T))
ggsave('plots/auc/avgpt_auc.png', plot=roc.plot(rocPtA, title='Mean per Patient (excl. final)', T, 0.3), width=6, height=6, units="in", limitsize=F, dpi = 300)

multi.roc.plot(list(rocPtM, rocPtA), threshold = 0.3) + labs(title='Per-Patient ROCs', subtitle = 'threshold = 0.3')
```

### Compare ROCs

```{r echo=F, fig.height=8, fig.width=6}
#get.auc.at(roc, p=c(1,0.8))
# get.auc.at(p53roc)
# get.auc.at(hgdpathroc)
# get.auc.at(pathroc)
# get.auc.at(roc)

roc$model = "CN Model"
m = bind_rows(
      get.roc.stat(p53roc),
      get.roc.stat(hgdpathroc),
      get.roc.stat(pathroc),
      get.roc.stat(roc, 0.3)) %>%
  dplyr::mutate(model = factor(model, levels=c('p53 IHC','HGD Only','HGD & LGD','CN Model'), ordered = T)) %>% 
  dplyr::mutate_if(is.numeric, list(~round(.,2))) 

cols = brewer.pal(4, "PRGn")
dodge = position_dodge(width=0.9)

ggplot(filter(m, Measure == 'AUC'), aes(model, value)) +
  geom_bar(stat='identity',position=dodge, fill='#7B3294') + geom_errorbar(aes(ymin=CI.min, ymax=CI.max), position=dodge, width=0.3, color='black' ) +
  geom_text(aes(label=paste(value,sep=''), y=CI.max), vjust=-0.5, position=dodge) + plot.theme + 
  scale_y_continuous(limits=c(0,1)) +
  theme(legend.position = 'none', text = element_text(size=14), axis.text.x = element_text(angle=45, hjust=1)) + 
  labs(x='', y='AUC', title='ROC Comparison') 

comp = ggplot(m, aes(model, value, ymin=CI.min, ymax=CI.max, group=Measure)) + #facet_wrap(~Measure) +
  geom_bar(aes(fill=Measure), stat='identity',position=dodge) + 
  geom_errorbar(position=dodge, width=0.3, color='black' ) +
  geom_text(aes(label=paste(value,sep=''), y=CI.max), vjust=-0.5, position=dodge) + plot.theme + 
  scale_y_continuous(limits=c(0,1)) +
  scale_fill_brewer(palette='Set2') +
  theme(text = element_text(size=14), axis.text.x = element_text(angle=45, hjust=1)) + 
  labs(x='', y='AUC', title='ROC Comparison') 

ggsave('plots/auc/path_p53_cn_auc.png', plot=comp, width=6, height=5, units="in", limitsize=F, dpi = 300)
rm(m,cols,dodge,comp)
```

```{r}
roc$model = 'All samples model'
rocNOHGD$model = 'Excluding HGD/IMC'
rocNOLGD$model = 'Excluding LGD/HGD/IMC'

m = bind_rows(
      get.roc.stat(roc,0.3),
      get.roc.stat(rocNOHGD,0.3)
#      get.roc.stat(rocNOLGD)
      ) %>% dplyr::mutate(model = factor(model)) %>% dplyr::mutate_if(is.numeric, list(~round(.,2))) 


dodge = position_dodge(width=0.9)
p = ggplot(m, aes(model, value, ymin=CI.min, ymax=CI.max, group=Measure)) + 
  geom_bar(aes(fill=Measure), stat='identity',position=dodge) + 
  geom_errorbar(position=dodge, width=0.3, color='black' ) +
  geom_text(aes(label=paste(value,sep=''), y=CI.max), vjust=-0.5, position=dodge) + plot.theme + 
  scale_y_continuous(limits=c(0,1)) +
  scale_fill_brewer(palette='Set2') +
  theme(legend.position = 'right',text = element_text(size=12)) + 
  labs(x='', y='ROC', title='LOO model comparisons', subtitle = 'threshold = 0.3') 
auc.leg = get.legend(p)
p
ggsave('plots/auc/loo_model_roc.png', plot=p+theme(legend.position = 'none'), height=4, width=4, units="in", limitsize=F, dpi = 300)

rm(p,dodge)
```

```{r}
roc$model = "per Sample"

m = bind_rows(
      get.roc.stat(roc,0.3),
      get.roc.stat(rocEndoM,0.3),get.roc.stat(rocEndoA,0.3),
      get.roc.stat(rocPtM,0.3), get.roc.stat(rocPtA,0.3)
      #get.roc.stat(rocFirstEndo), get.roc.stat(rocLastEndo),
      #get.roc.stat(rocRandEndo)
      ) %>% 
  dplyr::mutate_if(is.numeric, list(~round(.,2))) %>%
  dplyr::mutate(model = factor(model, levels=c('per Sample','Mean per Endoscopy','Max per Endoscopy','Mean per Patient','Max per Patient')))
  
dodge = position_dodge(width=0.9)
p = ggplot(m, aes(model, value, ymin=CI.min, ymax=CI.max, group=Measure)) + 
  geom_bar(aes(fill=Measure), stat='identity',position=dodge) + 
  geom_errorbar(position=dodge, width=0.3, color='black' ) +
  geom_text(aes(label=paste(value,sep=''), y=CI.max), vjust=-0.5, position=dodge) + plot.theme + 
  scale_y_continuous(limits=c(0,1)) +
  scale_fill_brewer(palette='Set2') +
  theme(legend.position = 'bottom',text = element_text(size=14), axis.text.x = element_text(angle=45, hjust=1)) + 
  labs(x='', y='ROC ', title='Aggregated ROC Comparison', subtitle='threshold = 0.3') 
auc.leg = get.legend(p)
p

ggsave('plots/auc/multi_auc.png', plot=p+theme(legend.position = 'none'), height=6, width=8, units='in',dpi=300)
ggsave('plots/auc/roc_leg.png', plot=grid.arrange(auc.leg), height=1, width=4, unit='in', dpi=300)

rm(rocEndoM, rocEndoA, rocLastEndo, rocFirstEndo, rocRandEndo, rocNOHGD, rocNOLGD, rocPtM, rocPtA)
```


## Prediction probability distribution

```{r}
pg.samp = do.call(bind_rows,lapply(unique(pg.samp$Patient) %>% purrr::map( function(p) pg.samp %>% filter(Patient == p)), transform.by.time)) 
sq = c(0, seq(1,132,12),max(pg.samp$months.before.final))

pg.samp = pg.samp %>% mutate(
  brks = cut(months.before.final, breaks=sq, 
                 labels=c(c('Endpoint',apply(cbind(sq[-1], sq[-1]+11), 1, function(x) paste(x, collapse='-')))[-c(length(sq)-1,length(sq))], '>120'),include.lowest = T, ordered_result = T),
) %>% dplyr::mutate(brks = factor(brks, rev(levels(brks))))
```

The labels in each decile are the ratio of progessor samples that have probabilities within that region. The deciles that are greyed out show insigificant enrichment for the NP/P ratio.

```{r spacial, echo=F, warning=F, message=F, fig.width=8, fig.height=6}
maxP = 0.5; minP = 0.3

bind_rows(round(as_tibble(pROC::coords(roc, maxP, transpose=F)),2) %>% mutate(Risk = 'High'),
          round(as_tibble(pROC::coords(roc, minP, transpose=F)),2) %>% mutate(Risk = 'Low')) %>% 
  kable(caption='Sens/Spec at high/low thresholds') %>%
  kable_styling(full_width = F)
  
cuts = seq(0,1,0.1)
pg.samp = pg.samp %>% mutate(quants = cut(Prediction, breaks=cuts, include.lowest = T))
qt = pg.samp %>% group_by(quants, Status) %>% dplyr::summarise(n=length(Status) ) %>% spread(Status, n)

ft.fun<-function(NP,P) {
 f = chisq.test(rbind(cbind(NP,P),table(sum.patient.data$Status)))
 cbind.data.frame('p.value'=round(f$p.value, 4))
}

qt = left_join(qt, pg.samp %>% dplyr::group_by(quants) %>% dplyr::summarise ( 'mn'=mean(Prediction), 'sd'=sd(Prediction) ), by='quants')

pred.confidence = qt %>% dplyr::group_by(quants) %>% 
  dplyr::mutate( 'P.ratio'=P/sum(NP,P), 'p.value'=ft.fun(NP,P)$p.value, 'conf'=ifelse(p.value < 0.05, '*', '') ) %>% 
  separate(quants, c('r1','r2'), ',', remove = F) %>% 
  mutate_at(vars('r1','r2'), list(sub), pattern='\\[|\\]|\\(', replacement='') %>% mutate_at(vars(r1,r2), as.double) %>%  
  dplyr::mutate(Risk = ifelse(round(P.ratio,1)<.5, 'Low','High'))

pred.confidence = bind_cols(pred.confidence, 
                            data.frame(ci.low=qbeta(0.025, shape1=pred.confidence$P+.5, shape2 = pred.confidence$NP+.5),
                                       ci.high=qbeta(0.975, shape1=pred.confidence$P+.5, shape2 = pred.confidence$NP+.5)))

pcfit = lm(P.ratio~mn, data=pred.confidence)
rsid = rstandard(pcfit)
#pred.confidence[which( rsid >= 1 |  rsid <= -1 ), 'Risk'] = 'Moderate'
pred.confidence = pred.confidence %>% mutate(Risk = as.character(ifelse(r1 < maxP & r2> minP, 'Moderate', Risk)))

annotation.table = pred.confidence %>% dplyr::group_by(Risk) %>% 
  dplyr::summarise( 'n(NP)'=sum(NP), 'n(P)'=sum(P), 'p.value'=format.pval(mean(p.value), 3), 'min'=min(r1), 'max'=max(r2) ) %>%
  arrange(min)

pred.confidence$Risk = factor(pred.confidence$Risk, levels=c('Low','Moderate','High'), ordered = T)

pg.samp = left_join(pg.samp, pred.confidence %>% dplyr::select(quants, Risk), by='quants')

rm(pcfit, rsid)
```

```{r pcalib, echo=F, warning=F, message=F, fig.width=8, fig.height=6}
pcal = BarrettsProgressionRisk::showPredictionCalibration(pred.confidence %>% dplyr::rename(perc = P.ratio))
pcal

pred.confidence %>% dplyr::select(quants, NP,P,P.ratio,ci.low,ci.high,p.value,mn,Risk) %>% 
  set_names(c('Decile', 'NP samples', 'P samples', 'P:NP', 'CI Low','CI High', 'chi p-value', 'mean(Pr)', 'Risk class')) %>%
  write_tsv('plots/calibration_table.tsv')
ggsave(filename='plots/pred_calib.png', plot=pcal, width=4, height=4, units='in', dpi=300)

mini = pcal + labs(x='Predicted risk', y='Observed risk', title='') + theme(legend.position = 'none') + 
  scale_x_continuous(expand = c(0, 0), limits = c(-0.5, 1.5), breaks = cuts, labels = c(0,'','',0.3,'',0.5,'','','','',1)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(-0.5, 1.5), breaks = seq(0,1,0.25), labels = seq(0,1,0.25)) + 
  theme(axis.text = element_text(size=14))
mini
ggsave(filename='plots/pred_calib_mini.png', plot=mini, width=4, height=4, units='in', dpi=300) 

pred.confidence %>% dplyr::mutate_if(is.double, list(signif), 2) %>% kable() %>% kable_styling('striped')

per.endo = pg.samp %>% group_by(Hospital.Research.ID, Status, Patient, PID) %>% 
  dplyr::mutate('n.samples'=length(PID),'Samples'=paste(Samplename,collapse=',')) %>% 
  dplyr::summarise_at(vars(matches('Pred|RR|Year')), list(max)) %>% dplyr::select(-matches('^Sample$')) %>% 
  arrange(Patient, Endoscopy.Year, PID)  
per.endo = per.endo %>% dplyr::mutate(quants = cut(Prediction, breaks=cuts, include.lowest = T)) %>% 
  left_join(pred.confidence %>% dplyr::select(quants, Risk), by='quants') %>% 
  dplyr::select(-matches('quant'))# %>% dplyr::mutate(Risk = as.character(Risk))

pg.samp %>% group_by(Status, Risk) %>% dplyr::summarise(n = length(Risk)) %>% spread(Risk, n) %>% rowwise %>% 
  dplyr::mutate(n=sum(High,Low,Moderate), Low = Low/n, Moderate = Moderate/n, High = High/n) %>%
  mutate_if(is.double, list(round),2) %>%
  kableExtra::kable(caption='Training set per-sample predictions (ratios)') %>% kableExtra::kable_styling(full_width = F)

per.endo %>%  group_by(Status, Risk)  %>% dplyr::summarise(n = length(Risk)) %>% spread(Risk, n) %>% rowwise %>% 
  dplyr::mutate(n = sum(High,Low,Moderate), Low = Low/n, Moderate = Moderate/n, High = High/n) %>% mutate_if(is.double, list(round),2) %>%
  kableExtra::kable(caption='Training set per-endoscopy predictions') %>% kableExtra::kable_styling(full_width = F)

rm(per.endo)
```


```{r probhist, echo=F, warning=F, message=F, fig.width=5, fig.height=5}
myPal = BarrettsProgressionRisk::riskColors()[as.character(pred.confidence$Risk)]

ggplot(pg.samp, aes(Prediction)) + geom_histogram(aes(fill=..x..), color='white', breaks=cuts, show.legend = F) + 
  scale_fill_gradientn( colors=myPal) +
    geom_vline(xintercept=annotation.table$max, linetype='longdash', color='grey39') +
    geom_text(data=annotation.table %>% dplyr::group_by(Risk, min,max) %>% dplyr::summarise( 'samp.perc'= signif((sum(`n(NP)`, `n(P)`)/nrow(patient.info))*100, 3) ), aes(x=min+(max-min)/2, y=5, label=paste(samp.perc, '%',sep=''))) +
    plot.theme + labs(title='Sample predictions', y='n Samples', x='Probability') + theme(legend.position = 'none')

rm(annotation.table)
```

```{r relrisk, echo=F, fig.width=6, fig.height=6}
pg.samp %>% dplyr::select(Prediction,RR, quants, Risk)

#myPal = BarrettsProgressionRisk::riskColors()[levels(pred.confidence$Risk)[sort(c(pred.confidence$Risk, pred.confidence$Risk))]]

unadjRR = ggplot(pg.samp, aes(RR)) + 
  geom_histogram(aes(fill=..x..), color='white', bins=10, show.legend = F) +
  scale_fill_gradientn(colors = myPal,  name='log(RR)') + 
  labs(y='n Samples', x='log(Relative risk of cancer progression)', title='Unadjusted relative risk') + plot.theme + 
  theme(axis.text = element_text(size=12))
unadjRR

pg.samp %>% dplyr::select(Samplename,Prediction, RR, Risk) %>% arrange(RR) %>% 
  dplyr::rename('Probability' = 'Prediction', 'log(Relative Risk)' = 'RR', 'Risk class' = 'Risk') %>%
  write_tsv('plots/unadjusted_RR.tsv')

ggsave(filename='plots/unadj_relrisk.png', plot=unadjRR + labs(title='', x='', y=''), width=6, height=5, units='in', dpi=300)
rm(unadjRR, myPal)
```


### Mountain plots using spatial info

```{r, echo=F, warning=F, message=F, fig.height=10, fig.width=20}
min.theme = theme(text=element_text(size=8), panel.background=element_blank(), strip.background=element_rect(fill="grey97"), strip.text.y = element_text(size=6),
                   strip.text.x = element_text(size=12), axis.text = element_blank(), axis.ticks = element_blank(),
                   axis.line=element_line(color='black'), panel.grid.major.y =element_line(color='grey95'), panel.grid.major.x = element_blank(),
                   panel.border = element_rect(color="grey", fill=NA, size=0.5), panel.spacing.x = unit(0, 'lines'), panel.spacing.y = unit(0.1, 'lines')   ) 

setOgj <- function(nl) {
  if (nl >= 0 & nl < 2) return(1)
  if (nl >= 2 & nl < 4) return(2)
  if (nl >= 4 & nl < 7) return(3)
  if (nl >= 7) return(4)
}
pg.samp = pg.samp %>% mutate(ogj = factor(unlist(purrr::map(nl, setOgj)), ordered=T))

# early/late samples
df = left_join(globalCNdf, pg.samp[,c('Samplename','Endoscopy.Year', 'ogj', 'months.before.final', 'RR','Prediction')], by=c('variable'='Samplename')) %>% arrange(chr, start, Endoscopy.Year)

## count arms differently than segs??
df2 = df %>% dplyr::group_by(Status, Patient, Endoscopy.Year, months.before.final) %>% dplyr::summarise( 
  'max.Pred'=max(Prediction),
  'max.RR'=max(RR),
  'max.Path'=max(Pathology),
  '5MB windows'=length(which(seg.type != 'arms' & cn %in% c('loss','gain')))/length(unique(variable)),
  'Chromosome arms'=length(which(seg.type == 'arms' & cn %in% c('loss','gain')))/length(unique(variable)),
  '5MB ratio'=`5MB windows`/length(cn)
  ) %>% ungroup %>% mutate(Patient = factor(Patient))

df2 = df2 %>% dplyr::group_by(Status, Patient) %>% dplyr::mutate(
  'early'= months.before.final == max(months.before.final),   
  'late' = months.before.final == min(months.before.final)  
) %>% filter(n() > 1)

el = df2 %>% filter(early | late)
el2 = el %>% dplyr::group_by(Status, Patient) %>% dplyr::summarise( arms = abs(diff(`Chromosome arms`)), CNA = abs(diff(`5MB windows`)) )
wtA = signif(wilcox.test(filter(el2, Status == 'Progressor')$arms, filter(el2, Status == 'Non-Progressor')$arms)$p.value,1)

sl = df2 %>% dplyr::select(Status, Patient) %>% distinct() %>% group_by(Status) %>% tally() %>% spread(Status,n)
status.labels <- c(
  'Non-Progressor' = paste0('Non-Progressor\n(n=',sl$`Non-Progressor`,' patients)'),
  'Progressor' = paste0('Progressor\n(n=',sl$Progressor,' patients)')
)

filter(df2, early|late) %>% dplyr::mutate('Endoscopy' = case_when( early ~ 'Initial', TRUE ~ 'Final'  )) %>% 
  dplyr::select(Patient, Status, everything(), -early, -late, -Endoscopy.Year, -matches('5MB'), -months.before.final) %>%
  dplyr::rename('max(Pathology)' = 'max.Path', 'max(Pr)' = 'max.Pred', 'max(RR)' = 'max.RR') %>%
  write_tsv('plots/arms_initial_to_final.tsv')

p = ggplot(df2 %>% filter(early|late), aes(Endoscopy.Year, `Chromosome arms`, color=Patient)) + 
  facet_grid(~Status, labeller = labeller(Status=status.labels)) +
  geom_point() + geom_line(data=subset(df2, early | late)) + 
  theme(legend.position = 'none', text=element_text(size=11)) + 
  plot.theme + labs(x='Endoscopy Year', y='CN altered arms', title='Chromosome arms, initial vs final endoscopy', subtitle=paste('p-value=', wtA))
p
ggsave(filename='plots/status_hetero.png', plot=p, width=6.5, height=4, units='in', dpi=300)
rm(p,wtA)

wtA = signif(wilcox.test(subset(df2, early &  Status == 'Progressor')$`Chromosome arms`, subset(df2, early & Status == 'Non-Progressor')$`Chromosome arms`)$p.value, 3)
wtCN = signif(wilcox.test(subset(df2, early & Status == 'Progressor')$`5MB windows`, subset(df2, early & Status == 'Non-Progressor')$`5MB windows`)$p.value, 2)

wt.labels <- c(
  '5MB windows' = paste0('5MB windows\np-value=', wtCN),
  'Chromosome arms' = paste0('Chromosome arms\np-value=',wtA)
)

m = melt(filter(df2, early), measure.vars=c('5MB windows','Chromosome arms'), id.vars=c('Status'))
m$Status = factor(m$Status, ordered=F)

m %>% as_tibble() %>% dplyr::rename('CN count' = 'value', 'Segment type' = 'variable') %>% write_tsv('plots/initial_endo.tsv')

p = ggplot(m, aes(Status, value, fill=Status)) + 
  facet_wrap(~variable, scales='free_y', labeller = labeller(variable=wt.labels)) + 
  geom_jitter(width=0.2, color='grey39') + 
  geom_boxplot(outlier.colour = NA, alpha=0.8, width=0.3) +
  scale_x_discrete(breaks = as.character(unique(m$Status)), 
                   labels = c('Non-Progressor' = paste0('Non-Progressor\n(n=',sl$`Non-Progressor`,')'),'Progressor' = paste0('Progressor\n(n=',sl$`Progressor`,')'))) + 
  plot.theme + theme(legend.position = 'none', text=element_text(size=11)) +
  labs( y='CN altered count', title='Initial endoscopy')
p
ggsave(filename='plots/hetero_initial.png', plot=p, width=6, height=4, units='in', dpi=300)

ogjLimits = levels(factor(pg.samp$ogj, ordered = T))

rm(m,p,wtA,wtCN,wt.labels,el,el2,df,df2)
```


```{r, eval=F}
# Output every patient, only need to do this once when something changes
for (i in sort(unique(sum.patient.data$Patient))) {
  print(i)
  pt = filter(sum.patient.data, Patient == i)
  status = as.character(filter(sum.patient.data, Patient == i)$Status)

  pms = mtn.spatial(filter(globalCNdf, Patient == i & seg.type != 'arms'), pg.samp, ogjLimits, paste('Patient', pt, '(', status, ')' ))

  fl = paste0('patient_', i, '_',status,'.png')
  ggsave(paste('plots/spatial/', fl, sep=''), plot=grid.arrange(pms$grob), width=10.5, height=4.5, units='in', dpi=300)
}
```


```{r}
globalCNdf %>% filter(Patient %in% c(20, 51) & seg.type != 'arms') %>% 
  left_join(dplyr::select(pg.samp, 'Samplename','Endoscopy.Year','months.before.final','Block','ogj'), by=c('variable' = 'Samplename')) %>%
  dplyr::select(Patient, everything(), -chr.length, -seg.type, -Hospital.Research.ID, -Pathology, -p53.Status, -variable, -cn, -Endoscopy.Year, -Block) %>%
  dplyr::rename('Location'='locs', 'CN value'='value', 'Relative esophageal sample location (1-4)' = 'ogj', 'Months before final endoscopy' = 'months.before.final') %>%
  write_tsv('plots/spatial_pts_20_51.tsv')

p20 = mtn.spatial(globalCNdf %>% filter(Patient == 20 & seg.type != 'arms'), pg.samp, ogjLimits,axis.text.y=T )
ggsave(filename='plots/spatial/patient_20_P.png', plot=grid.arrange(p20$grob), width=16, height=6, units='in', dpi=300)

p51 = mtn.spatial(globalCNdf %>% filter(Patient == 51 & seg.type != 'arms'), pg.samp, ogjLimits)
ggsave(filename='plots/spatial/patient_51_NP.png', plot=grid.arrange(p51$grob), width=10.5, height=4.5, units='in', dpi=300)

grid.arrange(p20$grob, p51$grob, ncol=1)

pg.samp = pg.samp %>% dplyr::select(-ogj)

ggsave('plots/spatial/legend.png', plot=grid.arrange(p20$legend), width=2,height=2, units='in', dpi=300)

rm(p20, p51)
```



## How do the predictions look at each time point?

#### By pathology?
```{r, echo=F, warning=F, message=F, fig.height=10, fig.width=12}
#tmp = pg.samp %>% dplyr::mutate(Pathology = recode(Pathology, 'HGD' = 'HGD/IMC', 'IMC' = 'HGD/IMC'))
pg.samp %>% dplyr::group_by(Status, Pathology, Risk) %>% tally(name='Sample count') %>% ungroup %>% dplyr::rename('Risk class' = 'Risk') %>%
  write_tsv('plots/discovery_cohort_by_path.tsv')

rp = plot.risk.by.path(pg.samp %>% ungroup, label_bars = F, ret_df = F)

rp + labs(title='Discovery cohort samples', y='% samples classified', x='Sample pathology') + theme(strip.text = element_text(size=14))

ggsave(filename='plots/pred_by_path.png', plot=(rp + theme(legend.position = 'none') + labs(title='', x='',y='')), width=7.5, height=6.5, units='in', dpi=300)
rm(rp)
```

```{r p53IHC, echo=F, warning=F, message=F, fig.height=8, fig.width=8}
p53Path = pg.samp %>% dplyr::filter(Status == 'P' & !is.na(p53.Status)) %>% 
  dplyr::group_by(Pathology,p53.Status) %>% tally(name='Freq') %>% ungroup %>% 
  group_by(Pathology) %>% dplyr::mutate(nPath = sum(Freq), Status = 'p53 IHC', ratio = round(Freq/nPath,2))

p53Path$p53.Status = factor(p53Path$p53.Status, ordered = T)
p53Path = arrange(p53Path, Pathology, p53.Status)
col = brewer.pal(11, "PRGn")[c(10,2)]

p53Path %>% dplyr::select(-Freq,) %>% dplyr::rename('Sample count' = 'nPath', 'Sample ratio'='ratio') %>%
  dplyr::mutate(p53.Status = recode_factor(p53.Status,'0'='Normal', '1'='Aberrant'), Status = 'Progressor')  %>% write_tsv('plots/p53IHC.tsv')
  

rp = ggplot(p53Path, aes(Pathology, ratio*100, group=Pathology, fill=p53.Status)) + geom_bar(stat='identity', position = position_stack()) +
  scale_fill_manual(values=col, name='p53 IHC', labels=c('Normal','Aberrant')) + 
  geom_text(aes(label=paste('n=',nPath,sep=''), x=Pathology, y=101), nudge_y=1) +
  geom_text(aes(label=paste(ratio*100,'%',sep='')), position=position_stack(vjust = 0.5), size=5,color='white') +
  plot.theme + labs(title='p53 IHC in samples from progressive patients', y='% samples', x='Pathology') + 
  theme(legend.position = 'bottom', text = element_text(size=14))
rp
ggsave(filename='plots/p53_pred.png', plot=rp, width=6, height=5, units='in', dpi=300)

rm(p53Path, col, rp)
```

## Backwards from HGD

Clinicians always ask to see the timepoints peeled backwards to see how early prediction is possible.

```{r endos, warning=F, message=F, echo=F, fig.height=8, fig.width=6}
sq = c(0, 1, 1*12, 2*12, 4*12, 6*12, 8*12, 10*12, 15*12)

endos = pg.samp %>% filter(Status == 'P') %>% 
  dplyr::group_by(Status, months.before.final, brks, PID) %>% 
  dplyr::summarise('pred.endo'=ifelse(length(which(Risk == 'High')) > 0, 1, 0), 
              'n.endos'=length(unique(PID)), 
              'n.samples'=length(PID),
              'pred.samples'=length(which(Risk == 'High')),
              'max.path'=max(Pathology),
              'max.prediction' = max(Prediction)
              ) %>% ungroup %>% 
  dplyr::mutate(brks = cut(months.before.final, include.lowest = T, ordered_result = T,
     breaks=sq,  labels=c('Endpoint','1-12','12-24','24-48','48-72','72-96','96-120','>120'))) %>% 
  dplyr::mutate(brks = factor(brks, rev(levels(brks)))) 

endos = endos %>% mutate(brks = recode_factor(brks, '96-120' = '>96', '>120'='>96')) 

endos.m = endos %>% dplyr::filter(Status == 'P') %>% dplyr::group_by(brks) %>% 
  dplyr::summarise( 'pred.r'=sum(pred.endo)/sum(n.endos), 'n.samp'=sum(n.samples), 'n.endo'=sum(n.endos),'mean(P)'=mean(max.prediction),'sem(P)'=sd(max.prediction)/sqrt(length(brks)))

endos.m %>% set_names(c('Months', 'Ratio predicted endoscopies','Sample count', 'Endoscopy count', 'mean(Pr)', 'SEM(Pr)')) %>% write_tsv('plots/pred_endo.tsv')

rp = ggplot(endos.m, aes(x=brks, y=pred.r*100, ymin=(pred.r-`sem(P)`)*100, ymax=(pred.r+`sem(P)`)*100, group=brks)) +
  geom_bar(stat='identity', fill=BarrettsProgressionRisk:::riskColors()[['High']], show.legend = F) + 
  geom_errorbar(color='grey55', width=0.2) +
  labs(x='Months prior to endpoint', y='Sensitivity for high risk samples', title='High risk classifications prior to final endoscopy') + coord_cartesian(ylim = c(0,100)) + 
  plot.theme + theme( axis.text.x=element_text(angle=45,hjust=1), text=element_text(size=12) ) 

rp + geom_text(aes(label=paste(round(pred.r*100), '%', sep=''), y=(pred.r-`sem(P)`)*100), color='white', nudge_y=-3) +
  geom_text(aes(y=5, label=paste('n=',n.endo,'\ns=',n.samp,sep='')), color='white', nudge_y=2) + labs(subtitle='n = #endoscopies, s = #samples')

ggsave(filename='plots/pred_endo.png', plot=rp + labs(title='', x='', y=''), width=6, height=5, units='in', dpi=300)
rm(endos,endos.m,rp)
```

```{r pred_by_time}
pts = pg.samp %>% dplyr::group_by(Status, Patient, months.before.final, brks, PID) %>% 
  dplyr::select(Patient, Status, months.before.final, brks, PID, Block, Endoscopy.Year, Pathology, Prediction, RR, Risk) %>% ungroup %>% 
  mutate(Patient = factor(Patient)) %>% group_by(Patient) %>% arrange(Endoscopy.Year)

time.pred = pts %>% dplyr::group_by(Status, Patient, months.before.final, brks, Endoscopy.Year) %>% 
  dplyr::summarise(
              'endo.risk'= max(Risk),
              'n.endos'=length(unique(PID)), 
              'n.samples'=length(PID),
              'endo.path'= max(Pathology),
              'endo.prediction' = max(Prediction),
              'Endo.Year' = unique(Endoscopy.Year),
              ) %>% ungroup %>% 
  dplyr::mutate(brks = cut(months.before.final, include.lowest = T, ordered_result = T,
     breaks=sq,  labels=c('Endpoint','1-12','12-24','24-48','48-72','72-96','96-120','>120'))) %>% 
  dplyr::mutate(brks = factor(brks, rev(levels(brks)))) %>% 
  group_by(Patient) %>% filter(months.before.final == min(months.before.final) | months.before.final == max(months.before.final))
  

time.pred %<>% filter(Patient %in% (time.pred %>% group_by(Patient) %>% tally %>% filter(n>1) %>% dplyr::select(Patient) %>% pull))

time.diffs = time.pred %>% group_by(Status, Patient) %>% 
  arrange(Patient, Endoscopy.Year) %>%
  dplyr::summarise(
    min.prob = min(endo.prediction),
    max.prob = max(endo.prediction),
    prob.diff = abs(diff(endo.prediction)),
    time.diff = max(months.before.final),
    risk.diff = diff((as.integer(endo.risk))),
    Risk = case_when( risk.diff < 0 ~ min(endo.risk), TRUE ~ max(endo.risk) )
    )

risk.over.time = pg.samp %>% dplyr::mutate(Patient = factor(Patient)) %>%
  dplyr::group_by(Status, Patient, months.before.final, Endoscopy.Year) %>% 
  dplyr::summarise( 'n.samples'=length(PID),
              'max.Path'= max(Pathology),
              'max.Prob' = max(Prediction),
              'max.Risk'= max(Risk) ) %>%
  arrange(Patient, desc(months.before.final)) %>% dplyr::group_by(Patient) %>% 
  dplyr::mutate(months.since.initial = (Endoscopy.Year-min(Endoscopy.Year))*12,
                risk.diff = as.integer(max.Risk)-as.integer(max.Risk[which.min(months.since.initial)]),
                ie.risk.diff = case_when(
                  risk.diff[which.max(months.since.initial)] < 0 ~ max.Risk[which.min(months.since.initial)],
                  TRUE ~ max.Risk[which.max(months.since.initial)] ) ) %>% 
  ungroup %>% mutate_at(vars(risk.diff, ie.risk.diff), list(~factor(., ordered=T))) %>%
  mutate(risk.diff = dplyr::recode_factor(risk.diff,'-1' = 'Decreased Risk', '-2'='Decreased Risk','0'='No change', '1'='Increased Risk', '2'='Increased Risk', .ordered = T), Status = dplyr::recode_factor(Status, 'P'='Progressor', 'NP'='Non-Progressor'))

tb = risk.over.time %>% group_by(Patient, Endoscopy.Year) %>% tally() %>% filter(n > 1)

for (i in 1:nrow(tb)) {
  rows = which(risk.over.time$Patient == tb[i,]$Patient & risk.over.time$Endoscopy.Year == tb[i,]$Endoscopy.Year)
  risk.over.time[max(rows), 'months.since.initial'] = risk.over.time[max(rows), 'months.since.initial'] + 6
}

risk.over.time %>% dplyr::select(Patient, Status, max.Prob,max.Risk, max.Path, months.since.initial, -risk.diff, -ie.risk.diff, -Endoscopy.Year, -n.samples) %>% 
  set_names(c('Patient', 'Status', 'max(Pr)', 'max(Risk class)', 'max(Pathology)', 'Months since initial endoscopy')) %>%
  write_tsv('plots/risk_over_time.tsv')

p = ggplot(risk.over.time) + facet_wrap(~Status, scales='free', ncol = 2) +
  geom_line(aes(x=months.since.initial, y=reorder(Patient, desc(ie.risk.diff)), color=ie.risk.diff)) + 
  scale_color_manual(values=unlist(riskColors()), name='Initial:Final Risk') + 
  geom_point(aes(x=months.since.initial, y=reorder(Patient, desc(ie.risk.diff)), fill=max.Risk, shape=max.Path), color='white', size=4 ) +  
  #geom_point(data=filter(risk.over.time, max.Path %in% c('HGD','IMC')), aes(x=months.since.initial, y=reorder(Patient, desc(ie.risk.diff))), shape=8, color='white', size=4 ) +  
  scale_fill_manual(values=unlist(riskColors()), name='Max Risk',guide=guide_legend(override.aes=list(fill=unlist(riskColors()), color=unlist(riskColors())))) + 
  scale_shape_manual(values=c(21,22,23,24,25), name='Max Pathology',guide=guide_legend(override.aes=list(fill='grey39', color='white'))) +
  labs(x='Months since initial endoscopy', y='Patient', title='Risks over time per endoscopy') + 
  plot.theme + theme(text=element_text(size=16), panel.spacing.x=unit(1,'lines')  )
p

legend = get.legend(p)

ggsave(filename='plots/risks_over_time.png', plot = p + labs(title='',x='',y='') + theme(legend.position='none'), height=8, width=12, units='in')

ggsave(filename='plots/risks_over_time_legend.png', plot = grid.arrange(legend), height=5, width=2, units='in')

rm(p,sq,tb,pts,risk.over.time,time.pred)
```

```{r heatmaps, echo=F, warning=F, message=F, fig.width=5, fig.height=5 }
pg.samp = pg.samp %>% dplyr::mutate(nl = factor(nl, ordered=T), 
                                    Risk = factor(Risk, levels=c('Low','Moderate','High'), ordered = T), 
                                    Patient = factor(Patient, ordered=T), 
                                    Block = factor(Block, ordered=T))

pg.samp %>% filter(Patient == 69) %>% dplyr::select(Patient, Status, Prediction, Risk, Pathology, nl, months.before.final) %>%
  dplyr::rename('Esophageal location'='nl', 'Months before final'='months.before.final', 'Probability'='Prediction', 'Risk class'='Risk') %>%
  write_tsv('plots/heatmaps/P/pt_69.tsv')

p = BarrettsProgressionRisk::patientRiskTilesPlot(pg.samp %>% filter(Patient == 6) %>% dplyr::select(Endoscopy.Year,months.before.final, Pathology,Block,Risk) %>% 
                                                    dplyr::rename(Endoscopy = Endoscopy.Year, GEJ.Distance = Block), 'months.before.final' ) 
tilesLegend = get.legend(p)
p + theme(legend.position = 'none')
ggsave('plots/heatmaps/legend_long.png',plot=grid.arrange(tilesLegend), width=10, height=2, units='in', dpi=300)
rm(p,tilesLegend)

dir.create('plots/heatmaps/P',recursive = T, showWarnings = F)
dir.create('plots/heatmaps/NP',recursive = T, showWarnings = F)
dir.create('plots/heatmaps/tiny/P',recursive = T, showWarnings = F)
dir.create('plots/heatmaps/tiny/NP',recursive = T, showWarnings = F)


pg.samp %>% dplyr::select(Patient,Status,Block,Risk,months.before.final) %>% arrange(Patient, months.before.final) %>%
  dplyr::rename('Risk class' = 'Risk', 'Esophageal location' = 'Block', 'Months before final' = 'months.before.final') %>%
  write_tsv('plots/pt_heatmaps.tsv')

pg.samp %>% dplyr::select(Patient,Status,Block,Risk,Pathology, months.before.final) %>% arrange(Patient, months.before.final) %>%
  dplyr::rename('Risk class' = 'Risk', 'Esophageal location' = 'Block', 'Months before final' = 'months.before.final') %>% arrange(Status) %>%
  write_tsv('plots/pt_heatmaps_with_path.tsv')

# Tiny heatmaps
plist = purrr::map(sum.patient.data$Patient, function(pt) {
  patient = pg.samp %>% filter(Patient == pt) %>% dplyr::select(Patient,Status,Endoscopy.Year,Block,Risk,months.before.final) %>% 
    dplyr::rename(Endoscopy = Endoscopy.Year, GEJ.Distance = Block) %>% ungroup

  p = BarrettsProgressionRisk::patientRiskTilesPlot(patient, 'months.before.final','rev') + 
    theme(legend.position = 'none',panel.border = element_rect(color="black", fill=NA, size=3),  axis.ticks = element_blank(), axis.text = element_blank(), plot.margin=grid::unit(c(0,0,0,0), "mm")) +
    labs(x='',y='',title='') 

  ht = ifelse(length(unique(patient$GEJ.Distance)) > 1, length(unique(patient$GEJ.Distance)), length(unique(patient$GEJ.Distance))+0.5)
  wd = ifelse(length(unique(patient$months.before.final)) > 1, length(unique(patient$months.before.final)), length(unique(patient$months.before.final))+0.5 )
  
  ggsave(paste0('plots/heatmaps/tiny/',unique(patient$Status),'/pt_',pt,'.png'), plot = p, height=ht, width=length(unique(patient$months.before.final)), units='in', dpi=150, limitsize=F) 

  patient = pg.samp %>% filter(Patient == pt) %>% dplyr::select(Patient,Status,Pathology,Endoscopy.Year,Block,Risk,months.before.final) %>% 
    dplyr::rename(Endoscopy = Endoscopy.Year, GEJ.Distance = Block) %>% ungroup

  if (pt == 69)
      patient = pg.samp %>% filter(Patient == pt) %>% dplyr::select(Patient,Status,Endoscopy.Year,Block,Risk,months.before.final) %>% 
    dplyr::rename(Endoscopy = Endoscopy.Year, GEJ.Distance = Block) %>% ungroup

  
    p = BarrettsProgressionRisk::patientRiskTilesPlot(patient, 'months.before.final','rev', 0.5) + theme(legend.position = 'none') + 
      labs(x='Months to endpoint',y='Location: EGJ->Mouth') + 
      theme(text = element_text(size = 12), plot.title = element_text(face = 'bold', size=16), axis.text=element_text(size = 10))
  
      h = (length(unique(patient$GEJ.Distance))+1)*0.5
  if (length(unique(patient$GEJ.Distance)) <= 1) h = (length(unique(patient$GEJ.Distance))+1)*0.8
  w = (length(unique(patient$months.before.final))+1)*0.5
  if (length(unique(patient$months.before.final)) <= 1) w = (length(unique(patient$months.before.final))+1)*0.8
    
    ggsave(paste0('plots/heatmaps/',unique(patient$Status),'/pt_',pt,'.png'), plot = p, height=h, width=w, units='in', dpi=300, limitsize=F) 

    return(p)
})
names(plist) = sum.patient.data$Patient

```

```{r, echo=F, warning=F, message=F, fig.width=5, fig.height=15}
table.path.risk<-function(df) {
  summary.t = as.data.frame.matrix(with(df, table(Pathology, Risk)))
  summary.t$Pathology = rownames(summary.t)
  
  summary.t = summary.t %>% dplyr::group_by(Pathology) %>% dplyr::mutate( 'High.r'=round(High/sum(High,Low,Moderate),3), 
                                                'Low.r'=round(Low/sum(High,Low,Moderate),3),
                                                'Moderate.r'=round(Moderate/sum(High,Low,Moderate),3))
  for (g in unique(df$Risk)) {
    for (i in 1:nrow(summary.t)) {
      summary.t[i,g] = paste(summary.t[i,g], ' (', summary.t[i,paste(g,'.r',sep='')]*100, '%)',sep='')
    }
  }
  summary.t[,c('Pathology',levels(df$Risk))]
}

summary.t = table.path.risk(pg.samp)
summary.t %>% kableExtra::kable() %>% kableExtra::kable_styling('striped', full_width = F)
write_tsv(summary.t, path = 'plots/risk_summary.tsv')
rm(summary.t)
```

```{r, echo=F, message=F, warning=F}
pg.samp = pg.samp %>% dplyr::group_by(Patient) %>% dplyr::mutate( 'pred.ratio'=round(sum(Prediction)/length(Prediction),2))

p1 = do.call('arrangeGrob', c(plist[as.character(filter(sum.patient.data,Status == 'P')$Patient)], top='Progressors'))
p2 = do.call('arrangeGrob', c(plist[as.character(filter(sum.patient.data,Status == 'NP')$Patient)], top='Non-Progressors'))

ggsave('plots/heatmaps/pred_heatmap.png', plot=grid.arrange(p2,p1,nrow=2), width=52, height=16, units="in", limitsize=F, dpi = 300)
ggsave('plots/heatmaps/pg_heatmap.png', plot=p1, width=52, height=24, units="in", limitsize=F, dpi = 300)
ggsave('plots/heatmaps/npg_heatmap.png', plot=p2, width=52, height=24, units="in", limitsize=F, dpi = 300)

rm(p1,p2, plist)
#ggsave('plots/risk_legend.png', plot=grid.arrange(riskLegend), width=2,height=4, dpi=300)
#ggsave('plots/risk_legend_long.png', plot=grid.arrange(riskLegendLong), width=8,height=2, dpi=300)
```

# How predictive are the strongest coefs?
```{r}
features = features %>% arrange(desc(hazards)) 

ftsamples = tibble(); ftpatients = tibble()
for (i in 1:nrow(features)) {
  ft = features %>% dplyr::slice(i)
  ft_cn = globalCNdf %>% filter(chr == ft$chr & start >= ft$start & end <= ft$end & 
                      seg.type == case_when((ft$end-ft$start) > 5e6 ~ 'arms', TRUE ~ '5MB') &
                      as.character(cn) == as.character(ft$gl))

  ftsamples = bind_rows(ftsamples, ft_cn %>% dplyr::group_by(Status) %>% tally() %>% spread(Status, n) %>% 
    dplyr::mutate('Sample counts for:' = paste0(ft$gl, ' of ', ft$label)) %>% 
    dplyr::select('Sample counts for:', everything()))

  ftpatients = bind_rows(ftpatients, ft_cn %>% dplyr::select(Patient, Status) %>% distinct %>% 
    group_by(Status) %>%  tally() %>% spread(Status, n) %>%
    dplyr::mutate('Patient counts for:' = paste0(ft$gl, ' of ', ft$label)) %>% 
    dplyr::select('Patient counts for:', everything()))
}

ftsamples = ftsamples %>% dplyr::mutate_all(list(~replace_na(.,0)))
ftpatients = ftpatients %>% dplyr::mutate_all(list(~replace_na(.,0)))


ftpatients %>% arrange(desc(Progressor))
ftsamples %>% arrange(desc(Progressor))


```


# Chr17

One way of looking at the p53 prediction (e.g. Adam Bass) is to see how predictive the p-arm loss in chr17 is.

## p arm
```{r, echo=T}
p11 = filter(globalCNdf, chr == 17 &  start == 1 & seg.type == 'arms')
p11Loss = p11 %>% filter(value < -1)
  
p11Loss %>% group_by(Status) %>% tally(name='p11 Loss') %>% 
  full_join( p11 %>% group_by(Status) %>% tally(name='Total Samples') ) %>% 
  kable(caption='Chr17 p11 loss') %>% kable_styling(full_width = F)

p11Loss %>% filter(value <= -1) %>% dplyr::select(Patient, Status) %>% distinct %>%
  group_by(Status) %>%  tally(name='p11 Loss') %>%
  kable(caption='Patients showing p53 loss & p53 IHC') %>% kable_styling(full_width = F)

p11Path = patient.info %>% filter(Samplename %in% p11Loss$variable) %>% dplyr::group_by(Pathology) %>% tally(name='samples')

p11Path %>% dplyr::mutate('ratio'=round(samples/sum(p11Path$samples),2)) %>% 
kable(caption='Ratio of all Chr17 p11 loss samples by pathology' ) %>% kable_styling(full_width = F)

p11Loss %>% group_by(variable) %>% tally

```

* Samples with a loss vs all samples
`r length(unique(p11Loss$variable))`:`r `length(unique(globalCNdf$variable))`,   `r round(length(unique(p11Loss$variable))/length(unique(globalCNdf$variable)),2)*100`%


* Samples with a loss vs all Progressor samples
Ratio of samples with p11 loss vs all progressor samples: `r length(unique(p11Loss$variable))`:`r `length(unique(filter(globalCNdf, Status == 'Progressor')$variable))`,   `r round(length(unique(p11Loss$variable))/length(unique(filter(globalCNdf, Status == 'Progressor')$variable)),2)*100`%

Ratio of progressor patients with a p11 loss in at least 1 sample: `r length(unique(p11Loss$Patient))`:`r length(subset(sum.patient.data, Status == 'P')$Patient)`r round(length(unique(p11Loss$Patient))/length(subset(sum.patient.data, Status == 'P')$Patient),2)*100`%

```{r}
patient.info %>% filter(Samplename %in% p11Loss$variable) %>% group_by(forcats::fct_explicit_na(p53.Status)) %>% tally %>% 
  set_names(c('p53 Status', 'n samples')) %>% kable(caption='p53 IHC status for samples with p11 loss') %>% kable_styling(full_width = F)

pi = patient.info
pi$p11Loss = as.numeric(pi$Samplename %in% p11Loss$variable)

pi %>% mutate(p53.Status = recode_factor(p53.Status, '1'='aberrant','0'='normal')) %>% dplyr::group_by(p11Loss, p53.Status) %>% tally %>% spread(p53.Status, n) 


```

17p11 loss and p53 IHC correlation  Fisher's `r broom::tidy(fisher.test(table(pi$p11Loss, pi$p53.Status))) %>% kable() %>% kable_styling(full_width=F)`

## q arm

```{r, echo=T}

q11 = filter(globalCNdf, chr == 17 & start > 1 & seg.type == 'arms')
q11gain = q11 %>% filter(value > 1)
  
q11gain %>% filter(value >= 1) %>% group_by(Status) %>% tally(name='q11 Gain') %>% 
  full_join( q11 %>% group_by(Status) %>% tally(name='Total Samples') ) %>% 
  kable(caption='Chr17 q11 gain') %>% kable_styling(full_width = F)

q11gain %>% filter(value >= 1) %>% dplyr::select(Patient, Status) %>% distinct %>%
  group_by(Status) %>%  tally(name='q11 Gain') %>%
  kable(caption='Patients showing q gain') %>% kable_styling(full_width = F)

q11Path = patient.info %>% filter(Samplename %in% q11gain$variable) %>% dplyr::group_by(Pathology) %>% tally(name='samples')

q11Path %>% dplyr::mutate('ratio'=round(samples/sum(q11Path$samples),2)) %>% 
kable(caption='Ratio of all Chr17 q11 gain samples by pathology' ) %>% kable_styling(full_width = F)

rm(q11, q11gain, q11Path)
rm(p11, p11Loss, p11Path, pi)
```

## Top Features

## Chr 17 Predictive 

One way of looking at the p53 prediction (e.g. Adam Bass) is to see how predictive the p-arm loss in chr17 is.
```{r, echo=F, warning=F, message=F, fig.height=5, fig.width=10, eval=T}
ccgenes = read_tsv('~/Data/CosmicCensusGenes.tsv', col_types = cols(.default = col_character())) %>% 
  filter(!grepl('^X|Y',`Genome Location`)) %>% 
  separate(`Genome Location`, c('chr','start','end'),':|-', remove=F) %>%
  dplyr::filter(start != '' | end != '') %>%
  dplyr::select(`Gene Symbol`, `Genome Location`, `Role in Cancer`, `Mutation Types`, chr, start, end)

# leave out the translocations
genes = ccgenes %>% dplyr::filter(grepl('TSG|oncogene', `Role in Cancer`) & `Mutation Types` != 'T') %>% 
  dplyr::select(`Gene Symbol`, `Genome Location`, chr, start, end) %>%
  dplyr::mutate(value = min(globalCNdf$value), Patient = '', chr = factor(chr,levels=levels(chr.lengths$chr),ordered=T), start = as.numeric(start), end=as.numeric(end)) %>% arrange(chr, start)
                          
genesGR = makeGRangesFromDataFrame(genes, keep.extra.columns = T)
chGR = makeGRangesFromDataFrame(features, keep.extra.columns = T)

globalCNdf %>% filter(chr == '17')

g17 = plot.cn.chr(globalCNdf, chrom='17', chr.lengths, haz=features, genes=genes, label=labeller(chr=label_both, Status=n.status)) + 
  labs(title="Chromosome 17", y='Relative CN') + scale_y_continuous(breaks = seq(-30,25,10), labels = c('-30', '','',0,'',20)) +
  theme(legend.position = 'none',  text=element_text(size=14))
g17

ggsave(filename='plots/hazards/chr17.png', plot=g17, width=8, height=4, units='in', dpi=300)

rm(g17)
```


```{r echo=F, warning=F, message=F, fig.height=5, fig.width=10, eval=F}
ov = findOverlaps(chGR, genesGR)
chGR$Genes = ''
for (hit in unique(queryHits(ov))) {
  chGR[hit]$Genes = paste(genesGR[subjectHits(ov[queryHits(ov) == hit])]$Gene.Symbol, collapse=',')
}

for (chr in unique(features$chr)) {  
  #print(chr)
  p = plot.cn.chr(globalCNdf,  chrom=chr, info=chr.lengths, haz=features, genes=genes, label=labeller(chr=label_both, Status=n.status)) +
    scale_y_continuous(breaks = seq(-30,25,10), labels = c('-30', '','',0,'',20)) +
    theme(strip.text.x = element_blank(), text=element_text(size=14)) + labs(title=paste('Chromosome',chr))
  f = paste('chr',chr,'.png',sep='')
  ggsave(paste('plots/hazards/',f,sep=''), plot=p, width=8, height=4, units='in', dpi=300)
}

annoG = ccgenes %>% filter(grepl('CDKN2A$|TP53|MYC$|SMAD4|VEG|ERBB2|EGFR|SMYD|ARID1A|FHIT|RUNX1', `Gene Symbol`)) %>% arrange(chr) %>%
  left_join(chr.lengths %>% dplyr::select(chrom,chr.length), by=c('chr'='chrom')) %>%
  mutate(ymin = round(range(globalCN.plot$data$mean.value))[1]/2, ymax = round(range(globalCN.plot$data$mean.value))[2]/2) %>%
  mutate(chr = factor(chr, levels=levels(chr.lengths$chr)), start = as.numeric(start), end = as.numeric(end))

p = globalCN.plot + #geom_rect(data=annoG, aes(xmin=start, xmax=end, ymin=ymin,ymax=ymax), color='red') +
  geom_point(data=annoG, aes(x=start, y=0), size=1.5, color='red') +
  geom_text_repel(data=annoG, aes(x=start, y=ymax/2, label=`Gene Symbol`), color='red', size=1.5 ) + theme(legend.position = 'none')

ggsave('plots/global_gene_ann.png', plot=p, width=10, height=5, units='in', dpi=300)

rm(ccgenes, annoG)
```

```{r, echo=F, message=F, warning=F, fig.height=5, fig.width=8, eval=F}
file = paste(cache.dir, 'chr17.predictive.Rdata', sep='/')
if (file.exists(file)) {
  message(paste("loading", file))
  load(file, verbose=T)
} else {
  stop(paste("File missing", file))
}

names(performance.at.1se)

mp = model.performance(performance.at.1se, coefs) 
mp$plot = mp$plot + labs(x='Models', title='Model performance for subsets of genomic regions') + 
  scale_colour_manual(values=c('grey39', 'grey39')) + theme(axis.text.x = element_text(angle=45,hjust=1)) +
  geom_hline(yintercept=mean(mp$performance$mean)+sd(mp$performance$mean), color='red', alpha=0.5) +
  geom_hline(yintercept=mean(mp$performance$mean)-sd(mp$performance$mean), color='red', alpha=0.5)
mp$plot
ggsave('plots/performance/performance_subsets.png', plot=mp$plot, height=5, width=9, units='in',dpi=300)


chr.info = read.table('hg19_genes.txt', sep='\t', header=T)
chr.info = chr.info[1:22,]

chr.info$info.content = chr.info$protein_coding/chr.info$length

# No correlation between information content (genes/length) and classification rate 
cor.test(mp$performance[1:22,'mean'], chr.info$info.content)


median(mp$performance$mean)
sd(mp$performance$mean)
mp$performance[17,'mean']

```

# Validation Cohort

```{r, echo=F}
pastefun<-function(x) {
  if ( !grepl('SLX-', x) ) x = paste0('SLX-',x)
  return(x)
}

patient.file = '~/Data/BarrettsProgressionRisk/QDNAseq/validation/sWGS_validation_batches.xlsx'

oac.info = readxl::read_xlsx(patient.file, 'OAC Test Samples') %>% 
  dplyr::mutate_at(vars(`SLX-ID`, `Block ID`), list(as.character)) %>% dplyr::filter(!is.na(`SLX-ID`)) %>%
  mutate(Samplename = paste(`SLX-ID`, `Index Sequence`, sep='.')) %>%
  dplyr::select(-matches('SLX|Index')) %>%
  mutate(Status = factor(Status, levels=levels(pg.samp$Status)), Block = factor(Block, levels=c('E','1','2'), ordered=T))

val.info = readxl::read_xlsx(patient.file, 'Final Validation Samples') %>% 
  dplyr::mutate_at(vars(`SLX-ID`, `Block ID`), list(as.character)) %>% dplyr::filter(!is.na(`SLX-ID`)) %>%
  mutate(Samplename = paste(`SLX-ID`, `Index Sequence`, sep='.')) %>%
  separate(`Block ID`, c('PID','Block'), sep='-|[:blank:]', remove=F) %>% dplyr::rename(PathID = `Block ID`) %>%
  dplyr::select(-matches('SLX|Index'), -`Path Notes`) %>%
  mutate(Pathology = recode(Pathology, 'GM'='NDBE'), Status = factor(Status, levels=levels(pg.samp$Status))) %>% 
  ungroup %>% group_by(`Hospital Research ID`) %>% 
  dplyr::mutate(`Weeks Follow Up` = as.double(difftime(`Last Endoscopy`, `Endoscopy`, units='weeks'))) 

val.demo = readxl::read_xlsx(patient.file, 'Demographics') %>% 
  mutate_at(vars(Status,Sex), list(~factor(.))) %>%
  mutate(Followup = as.double(difftime(`Last Endoscopy`, `Initial Diagnosis`, units='weeks'))/52) 

no_followup = filter(val.demo, Followup == 0 & Status == 'NP')
val.info = val.info %>% filter(Patient !=  no_followup$Patient)


## TODO clean this demo stuff up
val.demo = val.demo %>% mutate_at(vars(Sex,Smoking), list(~forcats::fct_explicit_na(.)))

val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status) %>% tally %>% spread(Status, n)
val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status,Sex) %>% tally %>% spread(Status, n)
chisq.test(table(val.demo$Sex, val.demo$Status))

val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status,Smoking) %>% tally %>% spread(Status, n) 
chisq.test(table(val.demo$Smoking, val.demo$Status)[-3,])


val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status) %>% dplyr::summarise_at(vars(Maximal), list(~mean(.,na.rm=T), ~sd(.,na.rm=T))) 

val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status) %>% dplyr::summarise_at(vars(`Age at diagnosis`), list(~mean(.,na.rm=T), ~sd(.,na.rm=T))) 

val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status) %>% dplyr::summarise_at(vars(Followup), list(~mean(.,na.rm=T), ~sd(.,na.rm=T))) 
val.demo %>% filter(Patient != no_followup$Patient) %>% group_by(Status) %>% dplyr::summarise_at(vars(Followup), list(~min(.,na.rm=T), ~max(.,na.rm=T))) 

val.info %>% mutate(sample.age = as.double(difftime(Sys.Date(), Endoscopy, units='weeks'))/52) %>% 
  group_by(Status) %>% dplyr::summarise_at(vars(sample.age), list(~min(.), ~max(.), ~median(.)))

wilcox.test(filter(val.demo, Status == 'NP')$`Age at diagnosis`, filter(val.demo, Status == 'P')$`Age at diagnosis`)
wilcox.test(filter(val.demo, Status == 'NP')$Maximal, filter(val.demo, Status == 'P')$Maximal)

fisher.test(table(val.demo$Status, val.demo$Sex))
fisher.test(table(val.demo$Status, val.demo$Smoking))

# samples
val.info %>% ungroup %>% dplyr::select(Patient, Status) %>% distinct %>% group_by(Status) %>% tally
val.info %>% group_by(Status) %>% tally

val.info$Block = sapply(strsplit(val.info$Block, ''), function(x) {
  if (is.na(x)) return(1)
  if (length(x) == 1) {
    if (!grepl('\\d', x)) return(recode(x, 'A'=1, 'B'=2, 'C'=3, 'D'=4, 'E'=5, 'F'=6,'G'=7 ))
    return(as.numeric(x))
  }
  if (length(x) > 1) {
    i = which(grepl('\\d', x))
    return(as.numeric(x[i]))
  }
})

bind_rows(val.info %>% dplyr::select(`Hospital Research ID`, Status) %>% distinct %>% group_by(Status) %>% tally %>% 
  spread(Status,n) %>% add_column(' ' = 'n Patients', .before=1),
val.info %>% group_by(Status) %>% tally %>% spread(Status,n) %>% add_column(' ' = 'n Samples', .before=1)) %>% 
  kable(caption='Validation cohort') %>% kable_styling(full_width = F)


vdir ='~/Data/BarrettsProgressionRisk/Analysis/validation/pcf_perPatient/50kb'
residuals = do.call(bind_rows, purrr::map(list.files(vdir,'residuals.tsv',recursive = T, full.names = T), function(f) read_tsv(f, col_types = 'ccddddl')))
failedQC = filter(residuals, varMAD_median > 0.008)

rm(residuals)
```

## Per sample (alpha = 0.9)

```{r, fig.width=8, fig.height=12}
vdir ='~/Data/BarrettsProgressionRisk/Analysis/validation/pcf_perPatient/50kb/predictions_5e6_all'

all.val.preds = do.call(bind_rows, purrr::map(list.files(vdir, 'predictions.tsv', full.names = T, recursive = T), function(f) {
    pd = read_tsv(f, col_types = 'cddcccDcccc') %>% dplyr::select(`Hospital Research ID`, Sample, Probability, `Relative Risk`)
    return(pd)
})) %>% mutate(quants = cut(Probability, breaks=cuts, include.lowest = T)) %>%
  left_join(pred.confidence %>% dplyr::select(quants, Risk), by='quants') %>%
  filter(`Hospital Research ID` != no_followup$`Hospital Research ID`)

oac.preds = all.val.preds %>% filter(`Hospital Research ID` %in% unique(oac.info$`Hospital Research ID`))

oac.preds %>% group_by(`Hospital Research ID`) %>% 
  dplyr::summarise( Probability = max(Probability), `Relative Risk` = max(`Relative Risk`), Risk = max(Risk)) %>%
  group_by(Risk) %>% tally %>% kable(caption='OAC samples used as a positive control') %>% kable_styling(full_width = F)

all.val.preds = all.val.preds %>% 
  filter( !Sample %in% failedQC$samplename & !`Hospital Research ID` %in% unique(oac.info$`Hospital Research ID`) )

val.preds = left_join(all.val.preds, val.info, by=c('Hospital Research ID', 'Sample'='Samplename')) %>% 
  dplyr::select(Patient, everything()) %>% 
  mutate(Pathology = factor(Pathology, levels=c('NDBE','ID','LGD'), ordered=T)) %>%
  mutate(Final.Endoscopy = `Last Endoscopy`, Endoscopy.Year = Endoscopy, Block = factor(Block, ordered=T))  

# samples passed QC
val.preds %>% group_by(Status) %>% tally
```  


```{r}
roc = pROC::roc(Status ~ Probability, data=val.preds, auc=T, ci=T, of='thresholds')
roc.plot(roc) + labs(title='Validation cohort')
pROC::ci.auc(roc$auc)
ggsave('plots/validation/validatation_roc.png', plot=roc.plot(roc) + labs(title='Validation cohort'), width=5, height=5, units='in',dpi=300)
rm(roc)

totals = val.preds %>% group_by(Status) %>% tally
val.counts = val.preds %>% 
    group_by(Status, Risk) %>% tally %>% spread(Risk,n) %>% ungroup %>%
    left_join(totals, by='Status') %>% mutate_if(is.numeric, funs(./n)) %>% dplyr::select(-n)

ggplot(melt(val.counts, id.vars=c('Status')), aes(variable, value, group=Status, fill=Status)) + ylim(0,1) + 
    geom_bar(stat='identity', position='dodge') + geom_text(aes(label=round(value,2)), position = position_dodge(width=1)) +
    labs(title='Validation risk (sample ratios per NP:P)', x='', y='ratio') + theme_minimal() 

rm(totals,val.counts)
```


## By path

```{r}
val.preds %>% dplyr::group_by(Status, Pathology, Risk) %>% tally(name='Sample count') %>% dplyr::rename('Risk class' = 'Risk') %>% write_tsv('plots/val_cohort_by_path.tsv')

plot.risk.by.path(val.preds, label_bars = T) + labs(title='Validation cohort samples', y='% samples classified', x='Sample pathology')  + theme(strip.text = element_text(size=12))

rp = plot.risk.by.path(val.preds, label_bars = F) + labs(title='Validation cohort samples', y='% samples classified', x='Sample pathology')  + theme(strip.text = element_text(size=12))
rp
ggsave(filename='plots/validation/val_pred_by_path.png', plot=rp + theme(legend.position = 'none') + labs(title='',x='',y=''), width=6, height=7, units='in', dpi=300)
rm(rp)

get.year<-function(x) {
  as.numeric(format(as.Date(x), '%Y'))
}

val.preds = dplyr::rename(val.preds, 'Samplename'='Sample')
val.preds = do.call(bind_rows, 
                    lapply(unique(val.preds$Patient) %>% map( function(p) val.preds %>% filter(Patient == p) %>% 
                                   mutate_at(vars(Endoscopy.Year, Final.Endoscopy), list(~get.year(.)) )
                                 ), transform.by.time) ) 

dir.create('plots/validation/heatmaps/P/tiny',recursive = T, showWarnings = F)
dir.create('plots/validation/heatmaps/NP/tiny',recursive = T, showWarnings = F)

for (pt in unique(val.preds$Patient)) {
  patient = val.preds %>% filter(Patient == pt) %>% 
    dplyr::select(Patient, Status, Endoscopy, months.before.final, Block,Risk) %>% 
    dplyr::rename(GEJ.Distance = Block)

    p = BarrettsProgressionRisk::patientRiskTilesPlot(patient, 'months.before.final','rev') + theme(legend.position = 'none') + labs(x='',y='',title='') + theme(axis.ticks = element_blank(), axis.text = element_blank(), plot.margin=grid::unit(c(0,0,0,0), "mm"))

    ht = ifelse(length(unique(patient$GEJ.Distance)) > 1, length(unique(patient$GEJ.Distance)), length(unique(patient$GEJ.Distance))+0.5)
    wd = ifelse(length(unique(patient$months.before.final)) > 1, length(unique(patient$months.before.final)), length(unique(patient$months.before.final))+0.5 )
  
    ggsave(paste0('plots/validation/heatmaps/',unique(patient$Status),'/tiny/pt_',pt,'.png'), plot = p, height=ht, width=length(unique(patient$months.before.final)), units='in', dpi=150, limitsize=F) 

    patient = val.preds %>% filter(Patient == pt) %>% 
      dplyr::select(Patient, Status, Endoscopy, months.before.final, Pathology,Block,Risk) %>% 
      dplyr::rename(GEJ.Distance = Block)

  p = BarrettsProgressionRisk::patientRiskTilesPlot(patient, 'months.before.final', 'rev', 0.5) + labs(x='months', y='location') + theme(legend.position = 'none',plot.title = element_text(face = 'bold', size=16), text = element_text(size=12), axis.text = element_text(size=12))

    h = (length(unique(patient$GEJ.Distance))+1)*0.8
  if (length(unique(patient$GEJ.Distance)) <= 1) h = (length(unique(patient$GEJ.Distance))+1)
  w = (length(unique(patient$months.before.final))+1)*0.8
  if (length(unique(patient$months.before.final)) <= 1) w = (length(unique(patient$months.before.final))+1)
    
    ggsave(paste0('plots/validation/heatmaps/',unique(patient$Status),'/pt_',pt,'.png'), plot = p, height=h, width=w, units='in', dpi=300, limitsize=F) 
}

sq = c(0, 1, 1*12, 2*12, 4*12, 6*12, 8*12, 10*12, 15*12)
val.preds = val.preds %>% mutate(months.before.final = ifelse(months.before.final > sq[length(sq)], sq[length(sq)], months.before.final))

v.prog = val.preds %>% filter(Status == 'P') %>% mutate(
  brks = cut(months.before.final, breaks=sq, 
                 labels=c(c('Endpoint',apply(cbind(sq[-1], sq[-1]+11), 1, function(x) paste(x, collapse='-')))[-c(length(sq)-1,length(sq))], '>120'), include.lowest = T, ordered_result = T),
) %>% dplyr::mutate(brks = factor(brks, rev(levels(brks))))

endos = v.prog %>% dplyr::group_by(Status, months.before.final, brks, PID) %>% 
  dplyr::summarise('pred.endo'=ifelse(length(which(Risk == 'High')) > 0, 1, 0), 
              'n.endos'=length(unique(PID)), 
              'n.samples'=length(PID),
              'pred.samples'=length(which(Risk == 'High')),
              'max.path'=max(Pathology),
              'max.prediction' = max(Probability)
              ) %>% ungroup %>% 
  dplyr::mutate(brks = cut(months.before.final, include.lowest = T, ordered_result = T, breaks=sq,  labels=c('Endpoint','1-12','12-24','24-48','48-72','72-96','96-120','>120'))) %>%
  dplyr::mutate(brks = factor(brks, rev(levels(brks)))) 

endos = endos %>% mutate(brks = recode_factor(brks, '96-120' = '>96', '>120'='>96')) 

endos.m = endos %>% dplyr::filter(Status == 'P') %>% dplyr::group_by(brks) %>% 
  dplyr::summarise( 'pred.r'=sum(pred.endo)/sum(n.endos), 'n.samp'=sum(n.samples), 'n.endo'=sum(n.endos),'mean(P)'=mean(max.prediction),'sem(P)'=sd(max.prediction)/sqrt(length(brks)))

rp = ggplot(endos.m, aes(x=brks, y=pred.r*100, ymin=(pred.r-`sem(P)`)*100, ymax=(pred.r+`sem(P)`)*100, group=brks)) +
  geom_bar(stat='identity', fill=BarrettsProgressionRisk:::riskColors()[['High']], show.legend = F) + 
  geom_errorbar(color='grey21', width=0.2) +
  geom_text(aes(y=3, label=paste(round(pred.r*100), '%', sep='')), color='white') +
  geom_text(aes(y=pred.r*100-5, label=paste('n=',n.endo,'\ns=',n.samp,sep='')), color='white') + 
  labs(x='Months prior to endpoint', y='Percent predicted', title='Predictions prior to final endoscopy', subtitle='n = #endoscopies, s = #samples') + coord_cartesian(ylim = c(0,100)) + 
  plot.theme + theme( axis.text.x=element_text(angle=45,hjust=1), text=element_text(size=12)) 
rp
ggsave(filename='plots/validation/pred_endo.png', plot=rp, width=6, height=5, units='in', dpi=300)

rm(rp,endos,endos.m,v.prog,sq)
```



```{r, fig.width=6, fig.height=8, eval=F}
# Risk Categories KM
require(survival)
require(survminer)
require(GGally)

ps = pg.samp %>% dplyr::select(Patient, Status,months.before.final, Risk) %>%
  group_by(Patient) %>% mutate(Diagnosis = as.integer(Status)-1)


sfit = survfit( Surv(months.before.final, Diagnosis, type='right') ~ Risk, data=ps, conf.int=0.95, conf.type='log', na.action=na.omit)
names(sfit$strata) = sub('.*=','', names(sfit$strata))

kmp = survminer::ggsurvplot(sfit, size=1, linetype="strata",risk.table=T, risk.table.col='strata',
                      conf.int=T, censor=T, legend='none', title='Risks over time', xlab='Time (months prior to final sample)',
                      ylab = 'Progression risk',
                      ggtheme=theme_minimal(), palette=unlist(BarrettsProgressionRisk::riskColors()), font.legend=c(14,'bold'))

kmp 

#ggsave('plots/risk_km.png', plot=print(kmp), height=9, width=8, units = 'in', dpi = 300)

```

```{r, eval=F}
vp = val.preds %>% dplyr::select(Patient, Status, months.before.final, Risk) %>% 
  mutate(Patient = as.factor(Patient)) %>% 
  group_by(Patient) %>% mutate(Diagnosis = as.integer(Status)-1) 

merged = bind_rows(ps,vp)

#merged = merged %>% mutate(Diagnosis = ifelse(Status == 'P' & months.before.final == 0, 1, 0))

#merged %>% 

sfit = survfit( Surv(months.before.final, Diagnosis) ~ Risk, data=merged, conf.int=0.95, conf.type='log', na.action=na.omit)
names(sfit$strata) = sub('.*=','', names(sfit$strata))

kmp = survminer::ggsurvplot(sfit, size=1, linetype="strata",risk.table=F, risk.table.col='strata',
                      conf.int=T, censor=T, legend='none',# xlim=c(0,150),
                      title='Risk Classifications', 
                      subtitle='Training & Validation cohorts',
                      xlab='Months prior to final sample',
                      ylab = 'Progression free fraction',
                      ggtheme=theme_minimal(), palette=unlist(BarrettsProgressionRisk::riskColors()), font.legend=c(14,'bold')) 

kmp$plot = kmp$plot + scale_x_reverse() 
kmp$table = kmp$table + scale_x_reverse()

kmp
ggsave('plots/risk_km.png', plot=print(kmp), height=8, width=7, units = 'in', dpi = 300)

```


# Clinical Story

## Vignettes

```{r, echo=F, warning=F, message=F, fig.width=5, fig.height=5}
sum.patient.data = sum.patient.data %>% mutate(smoking_str = recode(smoking, 'Y'='smoker', 'N'=''))
sum.patient.data$smoking_str = replace_na(sum.patient.data$smoking_str, '') 

create.desc<-function(pt,age,sex,smk,len,date) {
  str =  paste0('Patient ', pt, ': ', age, sex)
  if (smk != '' & !is.na(smk)) str = paste0(str, ', ', smk)
  str = paste0(str,'\n')
  
  if (len != '' & !is.na(len)) str = paste0(str, len,'cm length ')
  
  str = paste0(str, 'BE diagnosed ', date)
  
  return(str)
}
sum.patient.data$desc = sum.patient.data %>% group_by(Patient) %>% rowwise() %>% dplyr::summarise(desc = create.desc(Patient, age.diagnosis, gender, smoking_str, M, initial.endo))

riskCols = BarrettsProgressionRisk::riskColors()
pt = 34
p34 = vig.plot(df=filter(pg.samp, Patient == pt),cols=riskCols) + 
  labs(title = filter(sum.patient.data, Patient == pt)$desc) +
  theme(legend.position = 'bottom', text = element_text(size=12)) 

#vig.legend = get.legend(p34)
#ggsave('plots/vignette/legend_long.png', plot=grid.arrange(vig.legend), width=8, height=1)
ggsave(paste0('plots/vignette/vig_',pt,'.png'), plot=p34+theme(legend.position = 'none'), width=6.35, height=4.5,dpi=300)

pt = 3
p3 = vig.plot(df=filter(pg.samp, Patient == pt),cols=riskCols) + 
    labs(title = filter(sum.patient.data, Patient == pt)$desc) + 
    theme(legend.position = 'none', text = element_text(size=12))
ggsave(paste0('plots/vignette/vig_',pt,'.png'), plot=p3, width=6.35, height=4,dpi=300)

pt = 68
p68 = vig.plot(df=filter(pg.samp, Patient == pt),cols=riskCols) + 
    labs(title = filter(sum.patient.data, Patient == pt)$desc) + 
    theme(legend.position = 'none', text = element_text(size=12))
ggsave(paste0('plots/vignette/vig_',pt,'.png'), plot=p68, width=6.35, height=6,dpi=300)

rm(p3, p34, p68)

for (pt in c(56,20,86,23,33,49,51,6,25)) {
  length = sum.patient.data %>% ungroup %>% filter(Patient == pt) %>% dplyr::summarise(years = final.endo-initial.endo) %>% pull
  height = as.numeric(max( pg.samp %>% ungroup %>% filter(Patient == pt) %>% dplyr::select(Block) %>% pull ) ) + 1
  
  p = vig.plot(df=subset(pg.samp, Patient == pt),cols=riskCols) + 
    theme(legend.position = 'bottom') + labs(title = filter(sum.patient.data, Patient == pt)$desc) 
  vigLeg = get.legend(p)
  p = p + theme(legend.position = 'none', text = element_text(size=12))
  ggsave(paste0('plots/vignette/vig_',pt,'.png'), plot=p, width=length, height=height,dpi=300)
  rm(p,length,height)
}
ggsave('plots/vignette/legend.png', vigLeg, width=8, height=1, dpi=300)
rm(vigLeg)

```

### Recommendations

 1.	Immediate RFA: HGD diagnosis, two or more consecutive high risk predictions
 2.	Recheck 6-12 months: One high risk prediction or aberrant p53 IHC
 3. Recheck endoscopy 12-24 months: One or more moderate risk predictions
 4. Regular surveillance 3-5 years: Two or more consecutive low risk predictions


#### From the penultimate endoscopy
```{r, rx, echo=F, warning=F, message=T, fig.width=8, fig.height=5}
rx<-function(Patient, brks, Prediction, Pathology, Risk, p53.Status) {
  rules = as.data.frame(matrix(ncol=0, nrow=0))
  # Consecutive
  for (i in 1:length(brks)) {
    risks = table(Risk[i:(i+1)])
    p53 = table(p53.Status[i:(i+1)])
    rule = 'None'
    if ( risks['High'] == 2 | length(which(grepl('HGD|IMC', Pathology[i:(i+1)]))) > 0 ) {
      rule = 1
    } else if ( risks['High'] == 1 | p53['1'] > 0 ) {
      rule = 2
    } else if ( risks['Moderate'] > 0 ) {
      rule = 3
    } else if ( risks['Low'] == 2) {
      rule = 4
    }
  rules = rbind(rules, cbind( as.character(brks[i]), as.character(brks[(i+1)]), rule))
  }  
  colnames(rules) = c('t1','t2', 'rule')
  rules$Patient = Patient
  rules
}

pg.samp = pg.samp %>% dplyr::mutate(p53.Status = factor(p53.Status, levels=c(0,1), ordered=T), Probability = Prediction) 

max.rules = purrr::map(unique(pg.samp$Patient), function(pt) {
    preds = pg.samp %>% filter(Patient == pt) %>% arrange(Endoscopy.Year, months.before.final)

    endo = BarrettsProgressionRisk:::.setUpRxTablePerEndo(preds %>% dplyr::mutate(Endoscopy = months.before.final, Sample = 'Samplename'), 'max', verbose=F) %>%
      mutate(Endoscopy = as.integer(factor(Endoscopy*-1)))
    
    #preds = preds %>% group_by(Endoscopy.Year, months.before.final) %>% dplyr::summarise(Pathology = max(Pathology)) %>% 
    #  left_join(endo, by=c('Endoscopy.Year','months.before.final')) %>% dplyr::rename('Endoscopy' = 'Endoscopy.Year') %>% 
    #  dplyr::select(-Initial.Endoscopy, -Final.Endoscopy)
    
    if (nrow(endo) == 1) return(NA)
    
    rx = BarrettsProgressionRisk:::.apply.rules(endo, 'Endoscopy', pred.confidence,'numeric') 
    
    left_join(rx, endo %>% dplyr::select(Endoscopy, months.before.final, Endoscopy.Year, Probability), by=c('Time 1'='Endoscopy')) %>%
      mutate(`Time 2` = ifelse(months.before.final <= 0,'Endpoint',`Time 2`)) %>% 
      mutate(Patient = pt) %>% mutate_at( vars(matches('Rule|Time')), funs(as.character) ) %>%
    left_join((preds %>% group_by(Endoscopy.Year, months.before.final) %>% dplyr::summarise(Pathology = max(Pathology))), by=c('Endoscopy.Year','months.before.final')) %>%
    dplyr::select(Patient, months.before.final, Endoscopy.Year, Probability, everything()) 
})
names(max.rules) = unique(pg.samp$Patient)
max.rules = max.rules[which(sapply(max.rules, is.data.frame))]

progressors = max.rules[as.character(sum.patient.data %>% filter(Status == 'P') %>% dplyr::select(Patient) %>% pull)]
nonprog = max.rules[as.character(sum.patient.data %>% filter(Status == 'NP') %>% dplyr::select(Patient) %>% pull)]

pids = names(which(unlist(  sapply(progressors, function(x) { nrow(x) >= 1 & grepl('Endpoint', x[nrow(x),'Time 2']) }) )))
nids = names(which(unlist(  sapply(nonprog, function(x) { nrow(x) >= 1 & grepl('Endpoint', x[nrow(x),'Time 2']) }) )))

penult = bind_rows( do.call(bind_rows, progressors[pids]) %>% mutate(Status = 'P'),
                    do.call(bind_rows, nonprog[nids]) %>% mutate(Status = 'NP') ) 

penult = penult %>% filter(`Time 2` != 'Endpoint') %>% arrange(Patient, months.before.final) %>% 
  group_by(Patient, Status) %>% dplyr::mutate(gid = 1:n() ) %>% dplyr::filter(gid == 1) 

penult %>% dplyr::select(Patient, Status) %>% distinct %>% group_by(Status) %>% tally %>% 
  kable(caption='n patients with at least two consecutive endoscopies prior to endpoints') %>%
  kable_styling(full_width = F)

penult %>% filter(Status == 'P' & Rule %in% c(3,4)) %>% kable(caption='Prog patients that would be missed if we moved Moderate to Low risk') %>% kable_styling(full_width = F)

penult %>% filter(Status == 'NP' & Rule == 1) %>% kable(caption='NP patients for RFA') %>% kable_styling(full_width = F)

penult %>% group_by(Status, Rule) %>% tally %>% spread(Rule,n) %>% 
  kable(caption='Penult Endoscopy: # patients by rx rules based on the max (P) per endoscopy') %>% kable_styling(full_width=F)

penult = penult %>% group_by(Status, Rule) %>% tally %>% spread(Rule,n) %>% dplyr::rename(RFA = `1`, `6-12 month recheck` = `2`, `12-24 month recheck` = `3`, `3-5 year surveillance` = `4`)
p = melt(penult, id.vars = 'Status') %>% mutate(Status = recode(Status, NP = 'Non-Progressor' ,P = 'Progressor')) %>%
  ggplot(aes(variable, value, group = Status)) + 
  geom_col(aes(fill=Status), position = position_dodge()) + geom_text(aes(label=paste0('n=',value)), position = position_dodge(width = 1)) + coord_flip() +
  scale_fill_manual(values = c('#4575B4', '#A50026'), name='' ) + plot.theme + labs(x='', y='# Patients', title='Penultimate Endoscopy Rx') + theme(legend.position = 'bottom')
p
ggsave('plots/penult_endo_rx.png', plot=p, width=6, height=6, units='in', dpi=300)
rm(p)
```


#### From the second endoscopy

```{r}
pids = names(which(unlist(sapply(progressors, function(x) {
  nrow(x) > 1 & !grepl('Endpoint', x[2,'Time 2']) & !grepl('None', x[2,'Rule'])
}))))

nids = names(which(unlist(sapply(nonprog, function(x) {
  nrow(x) > 1 & !grepl('Endpoint', x[2,'Time 2']) & !grepl('None', x[2,'Rule'])
}))))

second = bind_rows( do.call(bind_rows, progressors[pids]) %>% mutate(Status = 'P'),
                    do.call(bind_rows, nonprog[nids]) %>% mutate(Status = 'NP') ) 

second = second %>% dplyr::group_by(Patient, Status) %>% arrange(`Time 2`) %>% 
  dplyr::mutate(gid = 1:n() ) %>% dplyr::filter(gid == 2) 

second %>% arrange(Status, Rx) %>% dplyr::select(Patient, Status, Pathology, everything(), -Endoscopy.Year, -matches('Time|Rule'), -gid) %>%
  dplyr::rename('max(Pr)' = Probability, 'Recommendation' = 'Rx', 'max(Pathology)' = 'Pathology', 'Months before endpoint' = 'months.before.final') %>% 
  write_tsv('plots/second_endo_rx.tsv')

second %>% dplyr::select(Patient, Status) %>% distinct %>% group_by(Status) %>% tally %>% 
  kable(caption='n patients with at least two years follow-up') %>%
  kable_styling(full_width = F)

second %>% group_by(Status, Rx) %>% tally %>% spread(Status,n) %>% 
  kable(caption='Second Endoscopy: # patients by rx rules based on the max (P) per endoscopy') %>% kable_styling(full_width=F)

p = second %>% group_by(Status, Rule) %>% tally %>% spread(Rule,n) %>% dplyr::rename(RFA = `1`, `6-12 month recheck` = `2`, `12-24 month recheck` = `3`, `3-5 year surveillance` = `4`)  %>%
  melt(id.vars = 'Status') %>% mutate(Status = recode(Status, NP = 'Non-Progressor' ,P = 'Progressor')) %>%
  ggplot(aes(variable, value, group = Status)) + scale_y_continuous(expand=c(0,0), limits=c(0,25)) +
  geom_col(aes(fill=Status), position = position_dodge()) + 
  coord_flip() +
  scale_fill_manual(values = c('#4575B4', '#A50026'), name='' ) + plot.theme + 
  labs(x='', y='Number of Patients', title='Second Endoscopy Rx') + theme(legend.position = 'bottom')
p + geom_text(aes(label=paste0('n=',value)), position = position_dodge(width = 1)) 

ggsave('plots/second_endo_rx.png', plot=p, width=6, height=6, units='in', dpi=300)
rm(p)

second %>% filter(Status == 'P' & Rule == '1') %>% ungroup %>% filter(months.before.final < 120) %>%
  dplyr::summarise_at(vars(months.before.final), list(~min(.), ~max(.), ~mean(.), ~median(.))) 

second %>% filter(Status == 'NP' & Rule == '1') %>% ungroup  %>%
  dplyr::summarise_at(vars(months.before.final), list(~min(.), ~max(.), ~mean(.), ~median(.))) 

```




# CoxHD with clinical factors
```{r, eval=F}
library(CoxHD)

time<-function(x) {
  for (i in 1:nrow(x)) {
    if (i == 1) {
      x[i,'tstart'] = 0
      x[i, 'tstop'] = x[(i+1), 'Endoscopy.Year'] - x[i, 'Endoscopy.Year']
      if (is.na(x[i,'tstop']) || x[i,'tstop'] == 0) x[i,'tstop'] = 0.5
      next
    }  

    x[i,'tstart'] = x[(i-1), 'tstop']

    # Final sample is different -- will leave it out later
    if (i == nrow(x)) {
      x[i, 'tstop'] = x[i,'tstart'] + 0.5
      break
    }

    x[i, 'tstop'] =  x[i,'tstart'] + x[(i+1), 'Endoscopy.Year'] - x[i, 'Endoscopy.Year']

    # Samples taken in the same year
    if ( x[i, 'Endoscopy.Year'] == x[(i+1), 'Endoscopy.Year'] )
      x[i,'tstop'] = x[i,'tstop'] + 0.5 
  }
  return(x[,c('tstart','tstop')])
}


coxDF = do.call(bind_rows, purrr::map(unique(pg.samp$Patient), function(pt) {
  df = pg.samp %>% filter(Patient == pt) %>% arrange(Endoscopy.Year)
  df %>% bind_cols(time(df))
})) %>%  dplyr::select(Samplename, Patient, Status, Risk, Prediction, Pathology, tstart, tstop, Age.at.diagnosis, Smoking, Sex, Maximal) %>% 
  mutate(Diagnosis = as.integer(Pathology %in% c('HGD','IMC')) ) %>%  dplyr::select(Samplename, Patient, Status, tstart, tstop, Diagnosis, everything() )

coxDF = bind_cols(coxDF, as_tibble(dysplasia.df[coxDF$Samplename,features$label])) %>% 
  rename_at( vars(matches('\\d+:')), list(~paste0('chr',gsub(':|-','.', .))) ) %>% 
  mutate(Smoking = as.integer(Smoking)-1, Sex = as.integer(Sex)-1, Pathology = as.integer(Pathology)-1) %>% ungroup

coxSM = CoxHD::StandardizeMagnitude(dplyr::select(coxDF, -matches('Samplename|Patient|Status|Risk|Prediction|Pathology|tstart|tstop|Diagnosis')))
coxSM = bind_cols(dplyr::select(coxDF, Samplename, tstart, tstop, Diagnosis), as_tibble(coxSM))

surv = with(coxSM, Surv(tstart, tstop, Diagnosis))

groups = factor(c(rep('clinical',4),c(rep('genomic', nrow(features)))))
rfx = CoxRFX(as.data.frame(coxSM[,-c(1:4)]), surv, groups, verbose=T, nu=1)
summary(rfx)

plot(rfx)

wt = WaldTest(rfx)
survConcordance(surv ~ predict(rfx))

wt$label = sub('\\.', '-', sub('\\.', ':', sub('chr', '', rownames(wt))))
wt = merge(wt, features[,c('label','hazards', 'coef','stability')], by='label', all=T) 

wt[is.na(wt$hazards), c('hazards', 'coef.y')] = 0
plot.waldtest(wt) + labs(x='RFX coef',y='GLM coef', title='RFX vs GLM coefficients', subtitle='Labeled points have a significant p-value in RFX (except BE/ID)')



#gcols = colnames(coxDF)
#cols = features$label


coxSM = cbind(coxDF[,c(1:3)], StandardizeMagnitude(coxDF[,-c(1:3)]))


```



# Check COX

```{r}
cox.cache.dir = '~/Data/BarrettsProgressionRisk/Analysis/cox_5e6_all/50kb/'
file = paste(cox.cache.dir, 'model_data.Rdata', sep='/')
load(file, verbose=T)

file = paste(cox.cache.dir, "all.pt.alpha.Rdata" , sep='/')
load(file, verbose=T)

do.call(grid.arrange, c(plots, nrow=2))

file = paste0(cox.cache.dir, '/loo_0.9.Rdata')
load(file, verbose=T)

rocCox = pROC::roc(Status ~ FittedRR, data=pg.samp, auc=T, ci=T, of='thresholds')
roc.plot(rocCox) + labs(title = 'Cox model ROC')

first = pg.samp %>% group_by(Patient) %>% arrange(Endoscopy.Year) %>% filter(Endoscopy.Year == Endoscopy.Year[1])
rocCox2 = pROC::roc(Status ~ FittedRR, data = first, auc = T, ci = T, of='thresholds')
roc.plot(rocCox2) + labs(title = 'Cox model ROC, first endo')

last = filter(pg.samp, Endoscopy.Year == Final.Endoscopy)
rocCox3 = pROC::roc(Status ~ FittedRR, data = last, auc = T, ci = T, of='thresholds')
roc.plot(rocCox3) + labs(title = 'Cox model ROC, last endo')

rocPath = pROC::roc(Status ~ Pathology, data=pg.samp, auc=T, ci=T, of='thresholds')
roc.plot(rocPath) + labs(title = 'Pathology status ROC')

cor.test(pg.samp$FittedRR, as.integer(pg.samp$Pathology))

```

# Coefficients by # patients

```{r, echo=F}
# SNP
snp.info = readxl::read_xlsx('~/Data/BarrettsProgressionRisk/Analysis/SNP/metadata_T1T2.xlsx') %>% 
  mutate( PID = paste(PatientID, `Timepoint Code`, sep='_'), Patient = as.numeric(PatientID), Hospital.Research.ID = as.character(Patient), Samplename = PID ) 

model.dir = '~/Data/BarrettsProgressionRisk/Analysis/dv_snp_model/50kb/'
load(list.files(model.dir, 'all.pt.alpha',recursive = T, full.names = T), verbose=F)

dvsnp.coefs = coefs[['0.9']]
dvsnp.perf = performance.at.1se[['0.9']]$mean
rm(dysplasia.df, plots, models,  performance.at.1se)

dvsnp.preds = purrr::map(list.files('~/Data/BarrettsProgressionRisk/Analysis/dv_snp_model/50kb/loo', full.names = T), function(f) read_tsv(f, col_types = 'ccccccdd') )

dvsnp.preds = do.call(bind_rows, dvsnp.preds)
dvsnp.patient = length(unique(dvsnp.preds$Patient))
dvsnp.roc = pROC::roc(Status ~ Probability, data=dvsnp.preds, ci=T, of='thresholds')

dvsnp.dv.auc = pROC::auc(Status ~ Probability, data = dvsnp.preds %>% filter(cohort != 'snp') )
dvsnp.snp.auc = pROC::auc(Status ~ Probability, data = dvsnp.preds %>% filter(cohort == 'snp') )

dvsnp.preds = dvsnp.preds %>% mutate(type = ifelse(cohort == 'snp', 'snp', 'sWGS')) 

dvsnp.aucs = 
  left_join(
  dvsnp.preds %>% dplyr::select(Patient, type) %>% distinct %>% group_by(type) %>% tally(name='# patients'),
  dvsnp.preds %>% dplyr::select(Patient, type) %>% group_by(type) %>% tally(name='# timepoint samples'), by='type') %>% 
  add_row( type = 'All (snp+sWGS)', `# patients` = dvsnp.preds %>% dplyr::select(Patient) %>% distinct %>% tally %>% pull, `# timepoint samples` = dvsnp.preds %>% dplyr::select(Patient) %>% tally %>% pull) %>% 
  arrange(type) %>% 
  add_column( auc = c(dvsnp.roc$auc, dvsnp.snp.auc, dvsnp.dv.auc) ) %>%
  add_column( ci.min.auc = c(min(pROC::ci.auc(dvsnp.roc$auc)), min(pROC::ci.auc(dvsnp.snp.auc)), min(pROC::ci.auc(dvsnp.dv.auc))) ) %>%
  add_column( ci.max.auc = c(max(pROC::ci.auc(dvsnp.roc$auc)), max(pROC::ci.auc(dvsnp.snp.auc)), max(pROC::ci.auc(dvsnp.dv.auc))) ) %>%
  mutate_at(vars(-matches('auc')), as.factor)

dv.snp.p = ggplot(dvsnp.aucs , aes(`# patients`, auc, group=type, fill=type)) + 
  geom_col() + 
  geom_errorbar(aes(ymin=ci.min.auc, ymax=ci.max.auc), width=0.2, color='grey39') +
  geom_text(aes(label=round(auc,2), y=0.05)) + 
  scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_fill_brewer(palette = 'Set2', name='') + 
  labs(title='sWGS (all) + SNP Model', y='AUC') + theme(legend.position = 'bottom') +
  plot.theme
dv.snp.p

ggsave(filename='plots/performance/training_snpdv.png', plot=dv.snp.p, width=6, height=10, units='in', dpi=300)

# discovery + validation
model.dir = '~/Data/BarrettsProgressionRisk/Analysis/models_5e6_all/50kb/'
load(list.files(model.dir, 'loo_0.9.Rdata',recursive = T, full.names = T))
d.n.patient = length(unique(pg.samp$Patient))
d.roc = pROC::roc(Status ~ Prediction, data = pg.samp, ci=T, of='thresholds')

load(list.files(model.dir, 'all.pt.alpha',recursive = T, full.names = T))
d.coefs = coefs[['0.9']]
d.perf = performance.at.1se[['0.9']]$mean
rm(dysplasia.df, plots, pg.samp, models, nzcoefs, fits, performance.at.1se)

all.dir = '~/Data/BarrettsProgressionRisk/Analysis/dv_model_5e6_all_dvz/50kb/'
load(list.files(all.dir, 'loo_0.9.Rdata',recursive = T, full.names = T))
dv.n.patient = length(unique(pg.samp$Patient))
dv.n.samples = length(unique(pg.samp$Samplename))
dv.roc = pROC::roc(Status ~ Probability, data = pg.samp, ci=T, of='thresholds')

load(list.files(all.dir, 'all.pt.alpha',recursive = T, full.names = T))
dv.coefs = coefs[['0.9']]
dv.perf = performance.at.1se[['0.9']]$mean
rm(dysplasia.df, plots, pg.samp, models, nzcoefs, cvs, performance.at.1se)
```



```{r, fig.width=12, fig.height=8}
cache.dir = '~/Data/BarrettsProgressionRisk/Analysis/models_5e6_all'

filenames = list.files(cache.dir, 'supp*')

threshold = 0.3
info = do.call(bind_rows, purrr::map(list.files(cache.dir, 'merged.Rdata', recursive = T, full.names = T), function(f) {
  load(f,verbose=F)
  left_join(model.info, do.call(bind_rows, purrr::map(rocs, function(roc) {
    add_column(auc=as.numeric(pROC::auc(roc)), pROC::coords(roc, threshold, transpose=F))  
  })), by='auc')
})) %>% mutate('Cohort' = 'Discovery (patient subsets)', Type = 'sWGS')

info = add_row(info, n.patients = d.n.patient, cv.performance = d.perf, n.coefs = nrow(d.coefs), n.samples = 773, auc = d.roc$auc,
               threshold = pROC::coords(d.roc, threshold, transpose=F)$threshold,
               specificity = pROC::coords(d.roc, threshold, transpose=F)$specificity,
               sensitivity = pROC::coords(d.roc, threshold, transpose=F)$sensitivity,
               Cohort = 'Discovery (all)', Type = 'sWGS') 

info = add_row(info, n.patients = dv.n.patient, cv.performance = dv.perf, n.coefs = nrow(dv.coefs), n.samples = dv.n.samples, auc = dv.roc$auc,
               threshold = pROC::coords(dv.roc, threshold, transpose=F)$threshold,
               specificity = pROC::coords(dv.roc, threshold, transpose=F)$specificity,
               sensitivity = pROC::coords(dv.roc, threshold, transpose=F)$sensitivity,
               Cohort = 'Discovery+Validation', Type = 'sWGS') 

info = add_row(info, n.patients = dvsnp.patient, cv.performance = dvsnp.perf, n.coefs = nrow(dvsnp.coefs), n.samples = nrow(dvsnp.preds), auc = dvsnp.roc$auc,
               threshold = pROC::coords(dvsnp.roc, threshold, transpose=F)$threshold,
               specificity = pROC::coords(dvsnp.roc, threshold, transpose=F)$specificity,
               sensitivity = pROC::coords(dvsnp.roc, threshold, transpose=F)$sensitivity,
               Cohort = 'sWGS (all) + SNP', Type = 'All') 

info = add_row(info, n.patients = filter(dvsnp.aucs, type == 'snp')$`# patients`, cv.performance = NA, n.coefs = NA, n.samples = NA, 
               auc = filter(dvsnp.aucs, type== 'snp')$auc,
               threshold = NA, specificity = NA, sensitivity = NA,
               Cohort = 'sWGS (all) + SNP', Type = 'SNP')

info = add_row(info, n.patients = filter(dvsnp.aucs, type == 'sWGS')$`# patients`, cv.performance = NA, n.coefs = NA, n.samples = NA, 
               auc = filter(dvsnp.aucs, type== 'sWGS')$auc,
               threshold = NA, specificity = NA, sensitivity = NA,
               Cohort = 'sWGS (all) + SNP', Type = 'sWGS')

sem<-function(x) {
  sd(x)/sqrt(10)
}
# 
# est = info  %>% dplyr::filter(!grepl('SNP', Cohort)) %>% 
#   dplyr::mutate(n.patients = as.numeric(n.patients)) %>%
#   group_by(n.patients, Cohort, Type) %>% 
#   dplyr::summarise_at(vars(-n.samples,-threshold), list(~mean(.,na.rm=T))) %>% 
#   arrange(n.patients) 
# 
# ggplot(est, aes(n.patients, auc)) + geom_point() + geom_smooth(method = 'lm') + xlim(20, 300) + ylim(0.6,1)
# ggplot(est, aes(n.patients, specificity)) + geom_point() + geom_smooth(method = 'lm') + xlim(20, 300) + ylim(0.6,1)
# ggplot(est, aes(n.patients, sensitivity)) + geom_point() + geom_smooth(method = 'lm') + xlim(20, 300) + ylim(0.6,1)
# 
# fit = lm(sensitivity~n.patients, data=est)
# predict(fit, data.frame(n.patients = c(164, 190, 200, 250, 300)), interval = 'confidence' )
# 
# fit = lm((1-specificity)~n.patients, data=est)
# predict(fit, data.frame(n.patients = c(190, 200, 250, 300)), interval = 'confidence' )
# 
# fit = lm(auc~n.patients, data=est)
# predict(fit, data.frame(n.patients = c(190, 200, 250, 300, 325)), interval = 'confidence' )
# 
# fit = lm(n.patients~auc, data=est)
# predict(fit, data.frame(auc = c(0.7, 0.89, 0.92, 0.95)), interval = 'confidence' )



x = info %>% group_by(n.patients, Cohort, Type) %>% dplyr::summarise_at(vars(-n.samples,-threshold), list(~mean(.), ~sem(.))) %>% as_tibble %>% 
  mutate(n.patients = factor(n.patients,levels=c( seq(28,88,10), 164, 413, 165, 248  ), ordered=T), Cohort = factor(Cohort, levels=c('Discovery (patient subsets)', 'Discovery (all)', 'Discovery+Validation', 'sWGS (all) + SNP'), ordered=T)) %>% arrange(n.patients) %>% dplyr::mutate(n.patients = paste0('n=',n.patients))

x %>% write_tsv('plots/exdatafig6.tsv')

p = ggplot( x %>% filter(!is.na(cv.performance_mean)), aes(x=n.patients, y=cv.performance_mean, group=n.patients, fill=Cohort)) + 
  facet_grid(~Cohort, scales = 'free_x', space='free_x') + 
  geom_col() + geom_errorbar(aes(ymin=cv.performance_mean-cv.performance_sem, ymax=cv.performance_mean+cv.performance_sem), width=0.2, color='grey39') +
  geom_text(aes(label=round(cv.performance_mean,2)), y=0.05) + 
  scale_y_continuous(expand=c(0,0), limits=c(0,0.8)) + 
  scale_fill_brewer(palette = 'Set2',name='') + 
  plot.theme + theme(legend.position = 'bottom', strip.text = element_blank(), panel.spacing = unit(0.5, 'lines')) +
  labs(title='Cross validation performance', y='', x='')

legend = cowplot::get_legend(p)
p = p + theme(legend.position = 'none')

p1 = ggplot(x %>% filter(!is.na(n.coefs_mean)), aes(x=n.patients, y=n.coefs_mean, group=n.patients, fill=Cohort)) + 
  facet_grid(~Cohort, scales = 'free_x', space='free_x') + 
  geom_col() + geom_errorbar(aes(ymin=n.coefs_mean-n.coefs_sem, ymax=n.coefs_mean+n.coefs_sem), width=0.2, color='grey39') +
  geom_text(aes(label=round(n.coefs_mean,2)), y=5) +
  scale_y_continuous(expand=c(0,0)) + scale_fill_brewer(palette = 'Set2') + 
  plot.theme + theme(legend.position = 'none', strip.text = element_blank(), panel.spacing = unit(0.5, 'lines')) +
  labs(title='# Coefficients',y='', x='')

p2 = ggplot(x, aes(x=n.patients, y=auc_mean, group=n.patients, fill=Cohort)) +
  facet_grid(~Cohort, scales = 'free_x', space='free_x') + 
  geom_col() + 
  geom_errorbar(aes(ymin=auc_mean-auc_sem, ymax=auc_mean+auc_sem), width=0.2, color='grey39') +
  geom_text(aes(label=round(auc_mean,2)), y=0.05) +
  geom_text( data=filter(x,Cohort == 'sWGS (all) + SNP'), aes(label=Type, y=0.1) ) +
  scale_y_continuous(expand=c(0,0), limits=c(0,1)) + scale_fill_brewer(palette = 'Set2') + 
  plot.theme + theme(legend.position = 'none', strip.text = element_blank(), panel.spacing = unit(0.5, 'lines')) +
  labs(title='AUC', y='', x='# patients')

grid.arrange(arrangeGrob(p,p1,p2,ncol=1), arrangeGrob(legend), heights=c(4,1))

ggsave(filename='plots/performance/training_n_pts.png',
       plot=grid.arrange(arrangeGrob(p,p1,p2,ncol=1), arrangeGrob(legend), heights=c(6,1)), width=8, height=12, units='in', dpi=300)

```



```{r,eval=F}
dv.cache.dir = '~/Data/BarrettsProgressionRisk/Analysis/dv_model_5e6_all_dvz/'
load(paste0(dv.cache.dir, '/50kb/loo_0.9.Rdata'), verbose=T)
dv.preds = pg.samp
rm(plots, performance.at.1se, coefs, nzcoefs, fits, pg.samp)

dv.preds %<>% dplyr::mutate(Pathology = recode(Pathology, BE = 'NDBE'), Risk = case_when(Probability <= 0.3 ~ 'Low', Probability >= 0.5 ~ 'High', TRUE ~ 'Moderate')) %>%
  mutate(Pathology = factor(Pathology, levels = c('NDBE','ID','LGD','HGD','IMC'), ordered=T))

p = plot.risk.by.path(dv.preds) + labs(subtitle='164 patient model (d+v)')

ggsave(filename='plots/pred_by_path_all.png', plot=p, width=7, height=6, units='in', dpi=300)

dv.preds %>% group_by(Pathology, Risk) %>% tally %>% spread(Pathology, n) 
```

